---
title: "Introduction to Bioconductor"
description: >
  An introduction to Bioconductor
author: "Stephanie Hicks"
output:
  rmarkdown::html_document:
   highlight: pygments
   toc: true
   toc_depth: 3
vignette: >
  %\VignetteIndexEntry{Bioconductor}
  %\VignetteEncoding[ut8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```


# Overview

## Key resources

- Workshop material: [pkgdown website](https://stephaniehicks.com/jhuquantneuro2022)
- Code: [GitHub](https://github.com/stephaniehicks/jhuquantneuro2022)

## Learning objectives

1. Describe what is the Bioconductor and how it's different than CRAN
2. Describe the package types in Bioconductor
3. Recognize and work with core Bioconductor objects including `SummarizedExperiment`
4. Be able to perform a basic differential expression analysis with bulk RNA-seq
5. Introduce the core Bioconductor object called `SingleCellExperiment`.


# Introduction to Bioconductor

## A brief history of Bioconductor

The Bioconductor project was **started in the Fall of 2001**, as an initiative for the collaborative creation of extensible software for computational biology and bioinformatics.

The **goal of the project** is to develop tools for the statistical analysis and comprehension of large datasets and technological artifacts in rigorously and robustly designed experiments.
Beyond statistical analyses, the interpretation of statistical results is supported by packages providing biological context, visualization, and reproducibility.

Over the years, software packages contributed to the Bioconductor project have reflected the evolution and emergence of several high-throughput technologies, from microarrays to single-cell genomics, through **many variations of sequencing experiments** (e.g., RNA-seq, ChIP-seq, DNA-seq), analyses (e.g., sequence variation, copy number variation, single nucleotide polymorphisms), and data modalities (e.g., flow cytometry, proteomics, microscropy and image analysis).

The Bioconductor project culminates at an **annual conference in North America in the summer**, while regional conferences offer great opportunities for networking in Europe, Asia, and North America.

The project is **committed to promote a diverse and inclusive community**, including a [Code of Conduct][bioc-code-of-conduct] enforced by a Code of Conduct committee.

![Timeline of major Bioconductor milestones alongside technological advancements.](https://github.com/carpentries-incubator/bioc-project/raw/main/fig/bioc-timeline.svg)

**Timeline of major Bioconductor milestones alongside technological advancements.**
Above the timeline, the figure marks the first occurrence of major events.
Within the timeline, the name of packages providing core infrastructure indicate the release date.
Below the timeline, major technological advancements contextualise the evolution of the Bioconductor project over time.


## A scientific project

- The original publication describes the aims and methods of the project at its inception is Gentleman et al. (2004).

- [Huber et al. (2015)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590) illustrates the progression of the project, including descriptions of core infrastructure and case studies, from the perspective of both users and developers.

- [Amezquita et al. (2020)](https://pubmed.ncbi.nlm.nih.gov/31792435) reviewed further developments of the project in the wake of single-cell genomics technologies.

Many more publications and book chapters cite the Bioconductor project, with recent example listed on the [Bioconductor website][bioc-publications].

In addition, there is a **core team** which is supported by an NIH grant, and
developers who contribute to the open source Bioconductor packages. 


## A package repository

## Overview and relationship to CRAN

Undoubtedly, **software packages are the best-known aspect of the Bioconductor project**.
Since its inception in 2001, the repository has grown over time to host thousands of packages.

The Bioconductor project has extended the preexisting CRAN repository -- much larger and general-purpose in scope -- to comprise R packages primarily catering for bioinformatics and computational biology analyses.


## The Bioconductor release cycle

The Bioconductor project also extended the packaging infrastructure of the CRAN repository to better support the deployment and management of packages at the user level.

In particular, the **Bioconductor projects features a 6-month release cycle** (typically around April and October), which sees a snapshot of the current version of all packages in the Bioconductor repository earmarked for a specific version of R.

**R itself is released on an annual basis** (typically around April), meaning that for each release of R, two compatible releases of Bioconductor packages are available.

- As such, **Bioconductor package developers are required to always use the version of R that will be associated with the next release of the Bioconductor** project.

- This means using the development version of R between October and April, and the release version of R between April and October.

Why? The strict Bioconductor release cycle **prevents users from installing temporally distant versions of packages** that were very likely never tested together.

This practice **reflects the development cycle of packages of both CRAN and Bioconductor**, where contemporaneous packages are regularly tested by automated systems to ensure that the latest software updates in package dependencies do not break downstream packages, or prompts those package maintainers to update their own software as a consequence.

Prior to each Bioconductor release, packages that do not pass the requires suites of automated tests are deprecated and subsequently removed from the repository.

This **ensures that each Bioconductor release** provides a suite of packages that are mutually compatible, traceable, and guaranteed to function for the associated version of R.

![Timeline of release dates for selected Bioconductor and R versions.](https://github.com/carpentries-incubator/bioc-project/raw/main/fig/bioc-release-cycle.svg)

**Timeline of release dates for selected Bioconductor and R versions.**
The upper section of the timeline indicates versions and approximate release dates for the R project.
The lower section of the timeline indicates versions and release dates for the Bioconductor project.
Source: [Bioconductor][bioc-release-dates].

## Package types

Packages are broadly divided in four major categories:

1. software
2. annotation data
3. experiment data
4. workflows

**Software packages** themselves can be subdivided into packages that 

  - provide **infrastructure** (i.e., classes) to store and access data
  - packages that provide **methodological tools** to process data stored in those data structures

This **separation of structure and analysis** is at the core of the Bioconductor project, encouraging developers of new methodological software packages to thoughtfully re-use existing data containers where possible, and **reducing the cognitive burden** imposed on users who can more easily experiment with alternative workflows without the need to learn and convert between different data structures.

**Annotation data packages** provide self-contained databases of diverse genomic annotations (e.g., gene identifiers, biological pathways).

- Different collections of annotation packages can be found in the Bioconductor project.
- They are identifiable by their respective naming pattern, and the information that they contain.

For instance, the so-called `OrgDb` packages (e.g., the `r BiocStyle::Biocpkg("org.Hs.eg.db")` package) provide information mapping different types of gene identifiers and pathway databases;
the so-called `EnsDb` (e.g., `r BiocStyle::Biocpkg("EnsDb.Hsapiens.v86")`) packages encapsulate individual versions of the Ensembl annotations in Bioconductor packages;
and the so-called `TxDb` packages (e.g., `r BiocStyle::Biocpkg("TxDb.Hsapiens.UCSC.hg38.knownGene")`) encapsulate individual versions UCSC gene annotation tables.

**Experiment data packages** provide self-contained datasets that are often used by software package developers to demonstrate the use of their package on well-known standard datasets in their **package vignettes**.

Finally, **workflow packages** exclusively provide collections of vignettes that demonstrate the combined usage of several other packages as a coherent workflow, but do not provide any new source code or functionality themselves.

## Online communication channels

### Support site

The Bioconductor [support site][bioc-support-site] provides a platform where users and developers can communicate freely (following the Bioconductor [Code of Conduct][bioc-code-of-conduct]) to discuss issues on a range of subjects, ranging from packages to conceptual questions about best practices.

### Slack workspace

The Bioconductor [Slack workspace][bioc-slack] is an open space that all community members are welcome to join (for free) and use for rapid interactions.
Currently, the "Pro" pricing plan kindly supported by core funding provides:

- Unlimited message archive
- Unlimited apps
- Group video calls with screen sharing
- Work securely with other organizations using Slack Connect

A wide range of channels have been created to discuss a range of subjects, and community members can freely join the discussion on those channels of create new ones to discuss new subjects.

Important announcements are posted on the `#general` channel.

### Developer Mailing List

The <bioc-devel@r-project.org> mailing list is used for communication between package developers, and announcements from the Biocondutor core team.

### A scientific and technical community

- [Scientific Advisory Board (SAB)][bioc-scientific-advisory-board] Meet Annually, External and Internal leader in the field who act as project advisors. No Term limits.
- [Technical Advisory Board (TAB)][bioc-technical-advisory-board]. Meet monthly to consider technical aspects of core infastructure and scientific direction of the project. 15 members, 3 year term. Annual open-to-all elections to rotate members. Current officers are Vince Carey (chair), Levi Waldron (vice Chair) Charlotte Soneson (Secretary).
- [Community Advisory Board (CAB)][bioc-community-advisory-board] Meet monthly to consider community outreach, events, education and training. 15 members, 3 year term. Annual open-to-all elections to rotate members. Current officers are Aedin Culhane (chair), Matt Ritchie (co Chair), Lori Kern (Secretary).
- [Code of Conduct committee][bioc-code-of-conduct]


# Getting started with Bioconductor

Before we get started, you need to know how to install Bioconductor packages.
The most important details are:

-   Bioconductor is a package repository, like CRAN
-   All Bioconductor packages **must** be installed following the instructions
    here: <https://bioconductor.org/install>
-   Bioconductor packages are linked in their versions, both to each other and
    to the version of R
-   Bioconductor's installation function will look up your version of R and give
    you the appropriate versions of Bioconductor packages
-   If you want the latest version of Bioconductor, you need to use the latest
    version of R.

How do you know if a package is a Bioconductor package? 

For one thing, you can
just **google the package name** and you will see either CRAN or Bioconductor as a
first result (packages must be in one or the other, they are not allowed to be
on both repositories). 

But also, you can use Bioconductor's installation
function to install any packages, even ones on CRAN. 

By the way, you can install
multiple packages at once by making a string vector:
`BiocManager::install(c("foo","bar"))`

## Working with Bioconductor objects

Bioconductor's infrastructure is **built up of object classes**. 

An example of a class is **GRanges** (stands for
"genomic ranges"), which is a way to specify a set of ranges in a particular
genome, e.g. from basepair 101 to basepair 200 on chromosome 1 of the human
genome (version 38).

What's an object? 

Well **everything in R is an object**, but usually when we talk about Bioconductor objects, we mean data structures containing many attributes, so more complex than a vector or matrix. 

And the objects have specific **methods** that help you either access the information in the object, run analyses on the object, plot the object, etc. 

Bioconductor also allows for **class hierarchy**, which means that you can define a class of object that inherits the structure and methods of a **superclass** on which it depends. This last point is mostly important for people who are developing new software for Bioconductor (maybe that's you!)

We will introduce the core Bioconductor objects here. 


### SummarizedExperiment (SE)

First, we will discuss one of the most important classes of object, which is
the *SummarizedExperiment*, or SE.

SEs have the structure:

-   a matrix of data, rows are genomic features, and columns are samples
-   a table of data about the samples (columns)
-   a table of data about the features (rows)

A diagram of this 3-part structure can be found
[here](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/figure/F2/).

In SE, the 3 parts of the object are called 1) `assay`, 2) `colData` and 3)
`rowData` or `rowRanges`.

**Note**: There was a class of object that came before the SE, called the **ExpressionSet**, which was used primarily to store **microarray** data. 

Here we will skip over the **ExpressionSet**, and just look at SEs.

It helps to start by making a small toy SE, to see how the pieces come together. 
(Often you won't make an SE manually, but it will be downloaded from an external source, or generated by a function that you call, e.g. the  `r BiocStyle::Biocpkg("tximeta")` or some other data loading function.)

```{r message=FALSE}
library(SummarizedExperiment)
col_data <- data.frame(sample=factor(1:6),
                       condition=factor(c("A","A","B","B","C","C")),
                       treated=factor(rep(0:1,3)))
col_data
```

An important aspect of SEs is that the rows can optionally correspond to
particular set of **GRanges**

- e.g. a row of an SE could give the number of RNA-seq reads that can be assigned to a particular gene, and the row could also have metadata in the 3rd slot including, e.g. location of the gene in the
genome. 

In this case, we use the `rowRanges` slot to specify the information.

If we don't have ranges, we can just put a table on the "side" of the SE by
specifying `rowData`.

I will show in the example though how to provide `rowRanges`. 

Let's **use the first 10 genes in the Ensembl database** for human. 

The following code loads a database, pulls out all the genes (as *GRanges*), removes extra "non-standard" chromosomes, and then subsets to the first 10 genes.

```{r message=FALSE}
library(EnsDb.Hsapiens.v86)
txdb <- EnsDb.Hsapiens.v86
g <- genes(txdb)
g <- keepStandardChromosomes(g, pruning.mode="coarse")
row_ranges <- g[1:10]
```

We will make up some simulated "expression" measurements, and then store these in the SE. 

I call `list` so I can name the matrix, otherwise it would not be named.

```{r}
exprs <- matrix(rnorm(6 * 10), ncol=6, nrow=10)
se <- SummarizedExperiment(assay = list("exprs" = exprs),
                           colData = col_data,
                           rowRanges = row_ranges)
se
```

We see this object has one named matrix. 
The object could have multiple matrices (as long as these are the same shape). 

In that case you could access the first with `assay` and in general by name, e.g. `assay(se, "exprs")` or equivalently `assays(se)[["exprs"]]` .

```{r}
assayNames(se)
```

Finally, if we wanted to add data onto the rows, for example, the score of a test on the matrix data, we use the metadata columns function, or `mcols`:

```{r}
mcols(se)$score <- rnorm(10)
mcols(se)
```

Adding data to the column metadata is even easier, we can just use `$`:

```{r}
se$librarySize <- runif(6,1e6,2e6)
colData(se)
```

### Using the ranges of a SE

How does this additional functionality of the `rowRanges` facilitate faster data
analysis?

Suppose we are working with another data set besides `se` and we find
a region of interest on chromsome 1. 

If we want to pull out the expression data for that region, we just ask for the subset of `se` that overlaps. 

First, we **build the query region**, and then use the **GRanges** function `overlapsAny()` within
single square brackets (like you would subset any matrix-like object):

```{r}
query <- GRanges("1", IRanges(25000,40000))
se_sub <- se[overlapsAny(se, query), ]
```

We could have equivalently used the shorthand code:

```{r}
se_sub <- se[se %over% query,]
```

We get just three ranges, and three rows of the SE:

```{r}
rowRanges(se_sub)
assay(se_sub)
```

Another useful property is that we know metadata about the chromosomes, and the version of the genome. 

- **Note**: If you were not yet aware, the basepair position of a given feature, say gene *XYZ*, will change between versions of the genome, as sequences are added or rearranged.

```{r}
seqinfo(se)
```



## Downloading SE data

Let's download a SE object from [recount2](https://jhubiostatistics.shinyapps.io/recount/), which performs a basic summarization of public data sets with gene expression data. 

This **dataset contains RNA-seq samples from human airway epithelial cell cultures**. 

The paper is [here](https://www.ncbi.nlm.nih.gov/pubmed/25706956). The
structure of the experiment was that, cell cultures from 6 asthmatic and 6
non-asthmatics donors were treated with viral infection or left untreated
(controls). 

So we have 2 samples (control or treated) for each of the 12 donors.

```{r}
library(here)

# tests if a directory named "data" exists locally
if(!dir.exists(here("data"))) { dir.create(here("data")) }

file <- "asthma.rda"
if (!file.exists(here("data", file))){
  url <- "http://duffel.rail.bio/recount/SRP046226/rse_gene.Rdata"
  download.file(url, here("data", file))
}
load(here("data", file))
```


We use a custom function to produce a matrix which a count of RNA fragments for
each gene (rows) and each sample (columns).

(Recount project calls these objects `rse` for **RangedSummarizedExperiment**,
meaning it has `rowRanges` information.)

```{r}
my_scale_counts <- function(rse_gene, round=TRUE) {
  cts <- assays(rse_gene)$counts
  # mapped_read_count includes the count for both reads of a pair
  paired <- ifelse(colData(rse_gene)$paired_end, 2, 1)
  x <- (colData(rse_gene)$mapped_read_count / paired) / colSums(cts)
  assays(rse_gene)$counts <- t(t(assays(rse_gene)$counts) * x)
  if (round) {
    assays(rse_gene)$counts <- round(assays(rse_gene)$counts)
  }
  rse_gene
}

rse <- my_scale_counts(rse_gene)
```

We can take a peek at the column data:

```{r}
colData(rse)[,1:6]
```

The information we are interested in is contained in the `characteristics` column (which is a character list).

```{r}
class(rse$characteristics)
rse$characteristics[1:3]
rse$characteristics[[1]]
```


We can pull out the 3 and 4 element using the `sapply` function and the square bracket function. 

I know this syntax looks a little funny, but it's really just saying, use the single square bracket, pull out the third element (or fourth element).

```{r}
rse$condition <- sapply(rse$characteristics, `[`, 3)
rse$treatment <- sapply(rse$characteristics, `[`, 4)
table(rse$condition, rse$treatment)
```

Let's see what the `rowRanges` of this experiment look like:

```{r}
rowRanges(rse)
seqinfo(rse)
```

The `rowRanges` here were determined by the quantification method that the
*recount2* authors used. 

We don't know what the genome is from the `seqinfo`, but we could look this up from the project website.

The following code I use to clean up the condition and treatment variables:

```{r message=FALSE}
library(magrittr)
rse$condition %<>% (function(x) {
  factor(sub("-",".", sub("disease state: (.*)","\\1",x) ))
  })
rse$treatment %<>% (function(x) factor(sub("treatment: (.*)","\\1",x)))
```

Now we have:

```{r}
table(rse$condition, rse$treatment)
```

## Visualizing count matrix data in a SE

Here we just use a transformation so that we can compute meaningful distances on count data (without a larger discussion on normalization). 

We build a **DESeqDataSet** and then specify the experimental design using a `~` and the variables that we expect to produce differences in the counts. 
(These variables are used to assess how much technical variability is in the data, but not used in the transformation function itself.)

```{r}
library(DESeq2)
dds <- DESeqDataSet(rse, ~condition + treatment)
```

We use this function, which implements a **variance stabilizing transformation**:

```{r}
vsd <- vst(dds)
```

We calculate the variance across all samples (on the transformed data):

```{r}
library(matrixStats)
rv <- rowVars(assay(vsd))
o <- order(rv, decreasing=TRUE)[1:100]
```

Finally, before plotting a heatmap, we **extract the covariates** that we want to annotated the top of the plot.

```{r}
anno_col <- as.data.frame(colData(vsd)[,c("condition","treatment")])
anno_col
```

This code pull out the top of the transformed data by variance, and adds an annotation to the top of the plot. 

By default the rows and columns will be clustered by Euclidean distance. See `?pheatmap` for more details on this function (it's a very detailed manual page).

```{r}
library(pheatmap)
pheatmap(assay(vsd)[o,],
         annotation_col=anno_col,
         show_rownames=FALSE, 
         show_colnames=FALSE)
```

We can also easily make a PCA plot with dedicated functions:

```{r}
plotPCA(vsd, intgroup="treatment")
```

# SingleCellExperiment

An example of a class that extends the SE is **SingleCellExperiment**. 
This is a special object type for looking at single cell data.

For more details, there is a free online book "Orchestrating Single Cell
Analysis With Bioconductor" produced by a group within the Bioconductor Project,
with lots of example analyses:
[OSCA](http://bioconductor.org/books/3.15/OSCA.basic/).

Here we show a quick example of how this object extends the SE.

```{r message=FALSE}
library(SingleCellExperiment)
sce <- as(rse, "SingleCellExperiment")
sce
```

There are special functions dedicated to scaling the samples (we will discuss
this technical aspect soon):

```{r}
library(scran)
sce <- computeSumFactors(sce)
sizeFactors(sce)
```

Similarly, dedicated functions for transformations:

```{r}
sce <- logNormCounts(sce)
assayNames(sce)
```

And dedicated functions and new slots for reduced dimensions:

```{r}
set.seed(1)
sce <- fixedPCA(sce, rank=5, subset.row=NULL)
reducedDimNames(sce)
```

We can manually get at the PCs:

```{r}
pca <- reducedDim(sce, "PCA")
plot(pca[,1:2])
```

But we can more easily use dedicated visualization functions:

```{r message=FALSE}
library(scater)
plotReducedDim(sce, "PCA", colour_by="treatment")
```

# Session Info

```{r}
sessionInfo()
```





[biocviews-site]: https://www.bioconductor.org/packages/release/BiocViews.html
[bioc-code-of-conduct]: https://www.bioconductor.org/about/code-of-conduct/
[bioc-release-dates]: https://bioconductor.org/about/release-announcements/
[bioc-slack]: https://bioc-community.herokuapp.com/
[bioc-support-site]: https://support.bioconductor.org/
[bioc-course-materials]: https://bioconductor.org/help/course-materials/
[bioc-publications]: https://www.bioconductor.org/help/publications/
[bioc-scientific-advisory-board]: https://bioconductor.org/about/scientific-advisory-board/
[bioc-technical-advisory-board]: https://bioconductor.org/about/technical-advisory-board/
[bioc-community-advisory-board]: https://bioconductor.org/about/community-advisory-board/
[bioc-youtube]: https://www.youtube.com/user/bioconductor
[bioc-books]: http://www.bioconductor.org/books/release/
[bioc-code-of-conduct]: https://www.bioconductor.org/about/code-of-conduct/
