<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of single-cell analysis in R/Bioconductor • jhuquantneuro2022</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of single-cell analysis in R/Bioconductor">
<meta property="og:description" content="Overview of single-cell analysis in R/Bioconductor
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jhuquantneuro2022</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-bioconductor.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/intro-single-cell.html">Overview of single-cell analysis in R/Bioconductor</a>
    </li>
    <li>
      <a href="../articles/data-viz.html">Data visualizations and the tidyverse</a>
    </li>
    <li>
      <a href="../articles/problem-set.html">Problem Set 2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://twitter.com/stephaniehicks" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stephaniehicks/jhuquantneuro2022" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of single-cell analysis in
R/Bioconductor</h1>
                        <h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            
      
      
      <div class="hidden name"><code>intro-single-cell.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="key-resources">Key resources<a class="anchor" aria-label="anchor" href="#key-resources"></a>
</h3>
<ul>
<li>Workshop material: <a href="https://stephaniehicks.com/jhuquantneuro2022">pkgdown
website</a>
</li>
<li>Code: <a href="https://github.com/stephaniehicks/jhuquantneuro2022" class="external-link">GitHub</a>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-1">Part 1<a class="anchor" aria-label="anchor" href="#part-1"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Understand how count matrices are created from single-cell
experimental platforms and protocols</li>
<li>Recognize which basic principles and concepts were transferred from
bulk to single-cell data analyses</li>
<li>Understand the key differences between bulk and single-cell
data</li>
<li>Define what is a “Unique Molecular Identifier”</li>
<li>Define multiplexing (and demultiplexing)</li>
<li>Overview of basics in single-cell data analyses including quality
control, normalization, feature selection, dimensionality reduction, and
clustering</li>
</ol>
</div>
<div class="section level3">
<h3 id="materials">Materials<a class="anchor" aria-label="anchor" href="#materials"></a>
</h3>
<p>We will go through the slides available here:</p>
<ul>
<li><a href="https://docs.google.com/presentation/d/1XHfN-5NCXtu1P4reotrckGplPXJBTlQraFpd8ViPf1g/edit?usp=sharing" class="external-link uri">https://docs.google.com/presentation/d/1XHfN-5NCXtu1P4reotrckGplPXJBTlQraFpd8ViPf1g/edit?usp=sharing</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="part-2">Part 2<a class="anchor" aria-label="anchor" href="#part-2"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives-1">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-1"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to create a single-cell count matrix and read it into R</li>
<li>Recognize and define the <code>SingleCellExperiment</code> S4 class
in R/Bioconductor to store single-cell data</li>
</ol>
</div>
<div class="section level3">
<h3 id="overview-1">Overview<a class="anchor" aria-label="anchor" href="#overview-1"></a>
</h3>
<p>NGS data from scRNA-seq experiments must be converted into a matrix
of expression values.</p>
<p>This is <strong>usually a count matrix</strong> containing the number
of reads (or UMIs) mapped to each gene (row) in each cell (column). Once
this quantification is complete, we can proceed with our downstream
statistical analyses in R.</p>
<p>Constructing a count matrix from raw scRNA-seq data requires some
thought as the term “single-cell RNA-seq” encompasses a variety of
different experimental protocols. This includes</p>
<ul>
<li>
<strong>droplet-based protocols</strong> like 10X Genomics, inDrop
and Drop-seq</li>
<li>
<strong>plate-based protocols</strong> with UMIs like CEL-seq(2) and
MARS-seq</li>
<li>
<strong>plate-based protocols</strong> with reads (mostly
Smart-seq2)</li>
<li>others like sciRNA-seq, etc</li>
</ul>
<p>Each approach requires a different processing pipeline to deal with
cell demultiplexing and UMI deduplication (if applicable). Here, we will
briefly describe some of the methods used to generate a count matrix and
read it into R.</p>
</div>
<div class="section level3">
<h3 id="creating-a-count-matrix">Creating a count matrix<a class="anchor" aria-label="anchor" href="#creating-a-count-matrix"></a>
</h3>
<p>As mentioned above, the exact procedure for quantifying expression
depends on the technology involved:</p>
<ul>
<li>For <strong>10X Genomics data</strong>, the <code>Cellranger</code>
software suite (Zheng et al. 2017) provides a custom pipeline to obtain
a count matrix. This uses STAR to align reads to the reference genome
and then counts the number of unique UMIs mapped to each gene.</li>
<li>Alternatively, pseudo-alignment methods such as <a href="https://doi.org/10.1186/s13059-019-1670-y" class="external-link"><code>alevin</code></a>
(Srivastava et al. 2019) can be used to obtain a count matrix from the
same data. This avoids the need for explicit alignment, which reduces
the compute time and memory usage.</li>
<li>For other <strong>highly multiplexed protocols</strong>, the <a href="http://bioconductor.org/packages/scPipe/" class="external-link"><code>scPipe</code></a>
package provides a more general pipeline for processing scRNA-seq data.
This uses the Rsubread aligner to align reads and then counts reads or
UMIs per gene.</li>
<li>For <strong>CEL-seq or CEL-seq2 data</strong>, the
<code>scruff</code> package provides a dedicated pipeline for
quantification.</li>
<li>For <strong>read-based protocols</strong>, we can generally re-use
the same pipelines for processing bulk RNA-seq data (e.g. Subread, RSEM,
salmon)</li>
<li>For <strong>any data involving spike-in transcripts</strong>, the
spike-in sequences should be included as part of the reference genome
during alignment and quantification.</li>
</ul>
<p>In all cases, the <strong>identity of the genes</strong> in the count
matrix should be <strong>defined with standard identifiers from Ensembl
or Entrez</strong>. These provide an unambiguous mapping between each
row of the matrix and the corresponding gene.</p>
<p>In contrast, <strong>a single gene symbol</strong> may be used by
multiple loci, or the mapping between symbols and genes <strong>may
change over time</strong>, e.g., if the gene is renamed.</p>
<p>This makes it difficult to re-use the count matrix as we cannot be
confident in the meaning of the symbols.</p>
<p>(Of course, <strong>identifiers can be easily converted to gene
symbols</strong> later on in the analysis. This is the recommended
approach as it allows us to document how the conversion was performed
and to backtrack to the stable identifiers if the symbols are
ambiguous.)</p>
</div>
<div class="section level3">
<h3 id="singlecellexperiment-class">
<code>SingleCellExperiment</code> Class<a class="anchor" aria-label="anchor" href="#singlecellexperiment-class"></a>
</h3>
<p>One of the <strong>main strengths</strong> of the Bioconductor
project lies in the use of <strong>a common data infrastructure</strong>
that powers interoperability across packages.</p>
<p>Users should be able to analyze their data using functions from
different Bioconductor packages without the need to convert between
formats.</p>
<p>To this end, the <code>SingleCellExperiment</code> class (from the
<code>SingleCellExperiment</code> package) serves as the common currency
for data exchange across 70+ single-cell-related Bioconductor
packages.</p>
<p>This class implements a data structure that stores all aspects of our
single-cell data - gene-by-cell expression data, per-cell metadata and
per-gene annotation - and manipulate them in a synchronized manner.</p>
<p><img src="figures/SingleCellExperiment.png" width="780px" style="display: block; margin: auto;"></p>
<p>[Amezquita et al. 2019 (<a href="https://doi.org/10.1101/590562" class="external-link uri">https://doi.org/10.1101/590562</a>)]</p>
<ul>
<li>Each piece of (meta)data in the <code>SingleCellExperiment</code> is
<strong>represented by a separate “slot”</strong>.</li>
</ul>
<p>(This terminology comes from the S4 class system, but that’s not
important right now).</p>
<p>If we imagine the <code>SingleCellExperiment</code> object to be a
cargo ship, the <strong>slots can be thought of as individual cargo
boxes with different contents</strong>, e.g., certain slots expect
numeric matrices whereas others may expect data frames.</p>
<p>If you want to know more about the available slots, their expected
formats, and how we can interact with them, check out this <a href="https://bioconductor.org/books/3.15/OSCA.intro/the-singlecellexperiment-class.html" class="external-link">chapter</a>.</p>
<div class="section level4">
<h4 id="singlecellexperiment-example">
<code>SingleCellExperiment</code> Example<a class="anchor" aria-label="anchor" href="#singlecellexperiment-example"></a>
</h4>
<p>Let’s show you what a <code>SingleCellExperiment</code> (or
<code>sce</code> for short) looks like.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 20006 3005 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(20006): Tspan12 Tshz1 ... mt-Rnr1 mt-Nd4l</span></span>
<span><span class="co">## rowData names(1): featureType</span></span>
<span><span class="co">## colnames(3005): 1772071015_C02 1772071017_G12 ... 1772066098_A12</span></span>
<span><span class="co">##   1772058148_F03</span></span>
<span><span class="co">## colData names(10): tissue group # ... level1class level2class</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: endogenous</span></span>
<span><span class="co">## altExpNames(2): ERCC repeat</span></span></code></pre>
<p>This <code>SingleCellExperiment</code> object has 20006 genes and
3005 cells.</p>
<p>We can pull out the counts matrix with the <code><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts()</a></code>
function and the corresponding <code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData()</a></code> and
<code><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData()</a></code>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##          1772071015_C02 1772071017_G12 1772071017_A05 1772071014_B06</span></span>
<span><span class="co">## Tspan12               0              0              0              3</span></span>
<span><span class="co">## Tshz1                 3              1              0              2</span></span>
<span><span class="co">## Fnbp1l                3              1              6              4</span></span>
<span><span class="co">## Adamts15              0              0              0              0</span></span>
<span><span class="co">## Cldn12                1              1              1              0</span></span>
<span><span class="co">##          1772067065_H06</span></span>
<span><span class="co">## Tspan12               0</span></span>
<span><span class="co">## Tshz1                 2</span></span>
<span><span class="co">## Fnbp1l                1</span></span>
<span><span class="co">## Adamts15              0</span></span>
<span><span class="co">## Cldn12                0</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 20006 rows and 1 column</span></span>
<span><span class="co">##          featureType</span></span>
<span><span class="co">##          &lt;character&gt;</span></span>
<span><span class="co">## Tspan12   endogenous</span></span>
<span><span class="co">## Tshz1     endogenous</span></span>
<span><span class="co">## Fnbp1l    endogenous</span></span>
<span><span class="co">## Adamts15  endogenous</span></span>
<span><span class="co">## Cldn12    endogenous</span></span>
<span><span class="co">## ...              ...</span></span>
<span><span class="co">## mt-Co2          mito</span></span>
<span><span class="co">## mt-Co1          mito</span></span>
<span><span class="co">## mt-Rnr2         mito</span></span>
<span><span class="co">## mt-Rnr1         mito</span></span>
<span><span class="co">## mt-Nd4l         mito</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 3005 rows and 10 columns</span></span>
<span><span class="co">##                        tissue   group # total mRNA mol      well       sex</span></span>
<span><span class="co">##                   &lt;character&gt; &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## 1772071015_C02       sscortex         1           1221         3         3</span></span>
<span><span class="co">## 1772071017_G12       sscortex         1           1231        95         1</span></span>
<span><span class="co">## 1772071017_A05       sscortex         1           1652        27         1</span></span>
<span><span class="co">## 1772071014_B06       sscortex         1           1696        37         3</span></span>
<span><span class="co">## 1772067065_H06       sscortex         1           1219        43         3</span></span>
<span><span class="co">## ...                       ...       ...            ...       ...       ...</span></span>
<span><span class="co">## 1772067059_B04 ca1hippocampus         9           1997        19         1</span></span>
<span><span class="co">## 1772066097_D04 ca1hippocampus         9           1415        21         1</span></span>
<span><span class="co">## 1772063068_D01       sscortex         9           1876        34         3</span></span>
<span><span class="co">## 1772066098_A12 ca1hippocampus         9           1546        88         1</span></span>
<span><span class="co">## 1772058148_F03       sscortex         9           1970        15         3</span></span>
<span><span class="co">##                      age  diameter        cell_id       level1class level2class</span></span>
<span><span class="co">##                &lt;numeric&gt; &lt;numeric&gt;    &lt;character&gt;       &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## 1772071015_C02         2         1 1772071015_C02      interneurons       Int10</span></span>
<span><span class="co">## 1772071017_G12         1       353 1772071017_G12      interneurons       Int10</span></span>
<span><span class="co">## 1772071017_A05         1        13 1772071017_A05      interneurons        Int6</span></span>
<span><span class="co">## 1772071014_B06         2        19 1772071014_B06      interneurons       Int10</span></span>
<span><span class="co">## 1772067065_H06         6        12 1772067065_H06      interneurons        Int9</span></span>
<span><span class="co">## ...                  ...       ...            ...               ...         ...</span></span>
<span><span class="co">## 1772067059_B04         4       382 1772067059_B04 endothelial-mural       Peric</span></span>
<span><span class="co">## 1772066097_D04         7        12 1772066097_D04 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772063068_D01         7       268 1772063068_D01 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772066098_A12         7       324 1772066098_A12 endothelial-mural        Vsmc</span></span>
<span><span class="co">## 1772058148_F03         7         6 1772058148_F03 endothelial-mural        Vsmc</span></span></code></pre>
</div>
</div>
</div>
<div class="section level2">
<h2 id="part-3">Part 3<a class="anchor" aria-label="anchor" href="#part-3"></a>
</h2>
<div class="section level3">
<h3 id="learning-objectives-2">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives-2"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Be able to describe a standard workflow for analyzing single-cell
data</li>
<li>Be able to run code for a standard workflow starting from loading a
<code>SingleCellExperiment</code> in R and identifying clusters.</li>
</ol>
</div>
<div class="section level3">
<h3 id="overview-2">Overview<a class="anchor" aria-label="anchor" href="#overview-2"></a>
</h3>
<p>Here, we provide an overview of the framework of a typical scRNA-seq
analysis workflow:</p>
<p><img src="figures/workflow.png" width="780px" style="display: block; margin: auto;"></p>
<p>In the simplest case, the workflow has the following form:</p>
<ol style="list-style-type: decimal">
<li>We compute <strong>quality control metrics</strong> to remove
low-quality cells that would interfere with downstream analyses. These
cells may have been damaged during processing or may not have been fully
captured by the sequencing protocol. Common metrics includes the total
counts per cell, the proportion of spike-in or mitochondrial reads and
the number of detected features.</li>
<li>We convert the counts into <strong>normalized expression
values</strong> to eliminate cell-specific biases (e.g., in capture
efficiency). This allows us to perform explicit comparisons across cells
in downstream steps like clustering. We also apply a transformation,
typically log, to adjust for the mean-variance relationship.</li>
<li>We perform <strong>feature selection to pick a subset of interesting
features</strong> for downstream analysis. This is done by modelling the
variance across cells for each gene and retaining genes that are highly
variable. The aim is to reduce computational overhead and noise from
uninteresting genes.</li>
<li>We apply <strong>dimensionality reduction to compact the
data</strong> and further reduce noise. Principal components analysis is
typically used to obtain an initial low-rank representation for more
computational work, followed by more aggressive methods like <span class="math inline">\(t\)</span>-stochastic neighbor embedding for
visualization purposes.</li>
<li>We <strong>cluster cells into groups</strong> according to
similarities in their (normalized) expression profiles. This aims to
obtain groupings that serve as empirical proxies for distinct biological
states. We typically interpret these groupings by identifying
differentially expressed marker genes between clusters.</li>
</ol>
</div>
<div class="section level3">
<h3 id="quick-start-simple">Quick start (simple)<a class="anchor" aria-label="anchor" href="#quick-start-simple"></a>
</h3>
<p>Here, we use the a droplet-based retina dataset from Macosko et
al. (2015), provided in the <code>scRNAseq</code> package. This starts
from a count matrix and finishes with clusters in preparation for
biological interpretation. We also demonstrate how to identify
differentially expressed genes between the clusters.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/MacoskoRetinaData.html" class="external-link">MacoskoRetinaData</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Quality control (using mitochondrial genes).</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="va">is.mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"^MT-"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">qcstats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html" class="external-link">perCellQCMetrics</a></span><span class="op">(</span><span class="va">sce</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="va">is.mito</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">filtered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/quickPerCellQC.html" class="external-link">quickPerCellQC</a></span><span class="op">(</span><span class="va">qcstats</span>, percent_subsets<span class="op">=</span><span class="st">"subsets_Mito_percent"</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="va">sce</span><span class="op">[</span>, <span class="op">!</span><span class="va">filtered</span><span class="op">$</span><span class="va">discard</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Normalization.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Feature selection.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="va">hvg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># PCA.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">sce</span>, ncomponents<span class="op">=</span><span class="fl">25</span>, subset_row<span class="op">=</span><span class="va">hvg</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Clustering.</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">bluster</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/colLabels.html" class="external-link">colLabels</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/clusterCells.html" class="external-link">clusterCells</a></span><span class="op">(</span><span class="va">sce</span>, use.dimred<span class="op">=</span><span class="st">'PCA'</span>,</span>
<span>    BLUSPARAM<span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/bluster/man/NNGraphParam-class.html" class="external-link">NNGraphParam</a></span><span class="op">(</span>cluster.fun<span class="op">=</span><span class="st">"louvain"</span><span class="op">)</span><span class="op">)</span>    </span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Visualization.</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">sce</span>, dimred <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html" class="external-link">plotUMAP</a></span><span class="op">(</span><span class="va">sce</span>, colour_by<span class="op">=</span><span class="st">"label"</span><span class="op">)</span></span></code></pre></div>
<div class="figure">
<img src="intro-single-cell_files/figure-html/unnamed-chunk-7-1.png" alt="UMAP plot of the retina dataset, where each point is a cell and is colored by the assigned cluster identity." width="700"><p class="caption">
UMAP plot of the retina dataset, where each point is a cell and is
colored by the assigned cluster identity.
</p>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Marker detection (between pairs of groups)</span></span>
<span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/findMarkers.html" class="external-link">findMarkers</a></span><span class="op">(</span><span class="va">sce</span>, test.type<span class="op">=</span><span class="st">"wilcox"</span>, direction<span class="op">=</span><span class="st">"up"</span>, lfc<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 13</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 24658 rows and 16 columns</span></span>
<span><span class="co">##                     Top      p.value          FDR summary.AUC     AUC.2</span></span>
<span><span class="co">##               &lt;integer&gt;    &lt;numeric&gt;    &lt;numeric&gt;   &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3                  1  0.00000e+00  0.00000e+00    0.867306  0.857315</span></span>
<span><span class="co">## TUBA1A                1  3.58784e-81  5.89793e-78    0.609195  0.479862</span></span>
<span><span class="co">## SNHG11                1  0.00000e+00  0.00000e+00    0.737419  0.747343</span></span>
<span><span class="co">## SYT1                  2 3.51850e-268 1.08449e-264    0.786614  0.454677</span></span>
<span><span class="co">## CALM1                 2  0.00000e+00  0.00000e+00    0.812019  0.665619</span></span>
<span><span class="co">## ...                 ...          ...          ...         ...       ...</span></span>
<span><span class="co">## VSIG1             24654            1            1           0         0</span></span>
<span><span class="co">## GM16390           24655            1            1           0         0</span></span>
<span><span class="co">## GM25207           24656            1            1           0         0</span></span>
<span><span class="co">## 1110059M19RIK     24657            1            1           0         0</span></span>
<span><span class="co">## GM20861           24658            1            1           0         0</span></span>
<span><span class="co">##                   AUC.3     AUC.4     AUC.5     AUC.6     AUC.7     AUC.8</span></span>
<span><span class="co">##               &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3          0.6814265  0.463403  0.806255  0.478814 0.1822596  0.888640</span></span>
<span><span class="co">## TUBA1A        0.1162662  0.315768  0.422655  0.289249 0.4529957  0.541010</span></span>
<span><span class="co">## SNHG11        0.0822847  0.562474  0.741261  0.703891 0.0228628  0.742098</span></span>
<span><span class="co">## SYT1          0.5354274  0.317921  0.786614  0.307464 0.0591646  0.549534</span></span>
<span><span class="co">## CALM1         0.1535742  0.370682  0.715928  0.425258 0.0105809  0.791806</span></span>
<span><span class="co">## ...                 ...       ...       ...       ...       ...       ...</span></span>
<span><span class="co">## VSIG1                 0         0         0         0         0         0</span></span>
<span><span class="co">## GM16390               0         0         0         0         0         0</span></span>
<span><span class="co">## GM25207               0         0         0         0         0         0</span></span>
<span><span class="co">## 1110059M19RIK         0         0         0         0         0         0</span></span>
<span><span class="co">## GM20861               0         0         0         0         0         0</span></span>
<span><span class="co">##                   AUC.9     AUC.10    AUC.11    AUC.12    AUC.13</span></span>
<span><span class="co">##               &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">## MEG3           0.850705 0.48570145  0.866916  0.873796  0.867306</span></span>
<span><span class="co">## TUBA1A         0.509171 0.33506617  0.552997  0.566389  0.609195</span></span>
<span><span class="co">## SNHG11         0.745361 0.74450392  0.737413  0.744927  0.737419</span></span>
<span><span class="co">## SYT1           0.823158 0.39904104  0.575948  0.584176  0.633476</span></span>
<span><span class="co">## CALM1          0.444978 0.00587859  0.749949  0.834002  0.812019</span></span>
<span><span class="co">## ...                 ...        ...       ...       ...       ...</span></span>
<span><span class="co">## VSIG1                 0          0         0         0         0</span></span>
<span><span class="co">## GM16390               0          0         0         0         0</span></span>
<span><span class="co">## GM25207               0          0         0         0         0</span></span>
<span><span class="co">## 1110059M19RIK         0          0         0         0         0</span></span>
<span><span class="co">## GM20861               0          0         0         0         0</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="quick-start-multiple-batches">Quick start (multiple batches)<a class="anchor" aria-label="anchor" href="#quick-start-multiple-batches"></a>
</h3>
<p>If you could like to see Quick Start for scRNA-seq data with multiple
batches, check out this workflow:</p>
<ul>
<li><a href="https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-multiple-batches" class="external-link uri">https://bioconductor.org/books/3.15/OSCA.intro/analysis-overview.html#quick-start-multiple-batches</a></li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin21.6.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.5.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /opt/homebrew/Cellar/openblas/0.3.21/lib/libopenblasp-r0.3.21.dylib</span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.2.1_2/lib/R/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] bluster_1.6.0               scran_1.24.0               </span></span>
<span><span class="co">##  [3] scRNAseq_2.10.0             scater_1.24.0              </span></span>
<span><span class="co">##  [5] ggplot2_3.3.6               scuttle_1.6.3              </span></span>
<span><span class="co">##  [7] SingleCellExperiment_1.18.0 SummarizedExperiment_1.26.1</span></span>
<span><span class="co">##  [9] Biobase_2.56.0              GenomicRanges_1.48.0       </span></span>
<span><span class="co">## [11] GenomeInfoDb_1.32.3         IRanges_2.30.1             </span></span>
<span><span class="co">## [13] S4Vectors_0.34.0            BiocGenerics_0.42.0        </span></span>
<span><span class="co">## [15] MatrixGenerics_1.8.1        matrixStats_0.62.0         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] AnnotationHub_3.4.0           BiocFileCache_2.4.0          </span></span>
<span><span class="co">##   [3] systemfonts_1.0.4             igraph_1.3.4                 </span></span>
<span><span class="co">##   [5] lazyeval_0.2.2                BiocParallel_1.30.3          </span></span>
<span><span class="co">##   [7] digest_0.6.29                 ensembldb_2.20.2             </span></span>
<span><span class="co">##   [9] htmltools_0.5.3               viridis_0.6.2                </span></span>
<span><span class="co">##  [11] fansi_1.0.3                   magrittr_2.0.3               </span></span>
<span><span class="co">##  [13] memoise_2.0.1                 ScaledMatrix_1.4.0           </span></span>
<span><span class="co">##  [15] cluster_2.1.4                 limma_3.52.2                 </span></span>
<span><span class="co">##  [17] Biostrings_2.64.1             pkgdown_2.0.6                </span></span>
<span><span class="co">##  [19] prettyunits_1.1.1             colorspace_2.0-3             </span></span>
<span><span class="co">##  [21] blob_1.2.3                    rappdirs_0.3.3               </span></span>
<span><span class="co">##  [23] ggrepel_0.9.1                 textshaping_0.3.6            </span></span>
<span><span class="co">##  [25] xfun_0.32                     dplyr_1.0.10                 </span></span>
<span><span class="co">##  [27] crayon_1.5.1                  RCurl_1.98-1.8               </span></span>
<span><span class="co">##  [29] jsonlite_1.8.0                glue_1.6.2                   </span></span>
<span><span class="co">##  [31] gtable_0.3.1                  zlibbioc_1.42.0              </span></span>
<span><span class="co">##  [33] XVector_0.36.0                DelayedArray_0.22.0          </span></span>
<span><span class="co">##  [35] BiocSingular_1.12.0           scales_1.2.1                 </span></span>
<span><span class="co">##  [37] edgeR_3.38.4                  DBI_1.1.3                    </span></span>
<span><span class="co">##  [39] Rcpp_1.0.9                    viridisLite_0.4.1            </span></span>
<span><span class="co">##  [41] xtable_1.8-4                  progress_1.2.2               </span></span>
<span><span class="co">##  [43] dqrng_0.3.0                   bit_4.0.4                    </span></span>
<span><span class="co">##  [45] rsvd_1.0.5                    metapod_1.4.0                </span></span>
<span><span class="co">##  [47] httr_1.4.4                    ellipsis_0.3.2               </span></span>
<span><span class="co">##  [49] farver_2.1.1                  pkgconfig_2.0.3              </span></span>
<span><span class="co">##  [51] XML_3.99-0.10                 uwot_0.1.14                  </span></span>
<span><span class="co">##  [53] sass_0.4.2                    dbplyr_2.2.1                 </span></span>
<span><span class="co">##  [55] locfit_1.5-9.6                utf8_1.2.2                   </span></span>
<span><span class="co">##  [57] labeling_0.4.2                tidyselect_1.1.2             </span></span>
<span><span class="co">##  [59] rlang_1.0.5                   later_1.3.0                  </span></span>
<span><span class="co">##  [61] AnnotationDbi_1.58.0          munsell_0.5.0                </span></span>
<span><span class="co">##  [63] BiocVersion_3.15.2            tools_4.2.1                  </span></span>
<span><span class="co">##  [65] cachem_1.0.6                  cli_3.3.0                    </span></span>
<span><span class="co">##  [67] generics_0.1.3                RSQLite_2.2.16               </span></span>
<span><span class="co">##  [69] ExperimentHub_2.4.0           evaluate_0.16                </span></span>
<span><span class="co">##  [71] stringr_1.4.1                 fastmap_1.1.0                </span></span>
<span><span class="co">##  [73] yaml_2.3.5                    ragg_1.2.2                   </span></span>
<span><span class="co">##  [75] knitr_1.40                    bit64_4.0.5                  </span></span>
<span><span class="co">##  [77] fs_1.5.2                      purrr_0.3.4                  </span></span>
<span><span class="co">##  [79] KEGGREST_1.36.3               AnnotationFilter_1.20.0      </span></span>
<span><span class="co">##  [81] sparseMatrixStats_1.8.0       mime_0.12                    </span></span>
<span><span class="co">##  [83] xml2_1.3.3                    biomaRt_2.52.0               </span></span>
<span><span class="co">##  [85] compiler_4.2.1                rstudioapi_0.14              </span></span>
<span><span class="co">##  [87] beeswarm_0.4.0                filelock_1.0.2               </span></span>
<span><span class="co">##  [89] curl_4.3.2                    png_0.1-7                    </span></span>
<span><span class="co">##  [91] interactiveDisplayBase_1.34.0 statmod_1.4.37               </span></span>
<span><span class="co">##  [93] tibble_3.1.8                  bslib_0.4.0                  </span></span>
<span><span class="co">##  [95] stringi_1.7.8                 highr_0.9                    </span></span>
<span><span class="co">##  [97] GenomicFeatures_1.48.3        desc_1.4.1                   </span></span>
<span><span class="co">##  [99] lattice_0.20-45               ProtGenerics_1.28.0          </span></span>
<span><span class="co">## [101] Matrix_1.4-1                  vctrs_0.4.1                  </span></span>
<span><span class="co">## [103] pillar_1.8.1                  lifecycle_1.0.1              </span></span>
<span><span class="co">## [105] BiocManager_1.30.18           jquerylib_0.1.4              </span></span>
<span><span class="co">## [107] RcppAnnoy_0.0.19              BiocNeighbors_1.14.0         </span></span>
<span><span class="co">## [109] bitops_1.0-7                  irlba_2.3.5                  </span></span>
<span><span class="co">## [111] httpuv_1.6.5                  rtracklayer_1.56.1           </span></span>
<span><span class="co">## [113] R6_2.5.1                      BiocIO_1.6.0                 </span></span>
<span><span class="co">## [115] promises_1.2.0.1              gridExtra_2.3                </span></span>
<span><span class="co">## [117] vipor_0.4.5                   codetools_0.2-18             </span></span>
<span><span class="co">## [119] assertthat_0.2.1              rprojroot_2.0.3              </span></span>
<span><span class="co">## [121] rjson_0.2.21                  withr_2.5.0                  </span></span>
<span><span class="co">## [123] GenomicAlignments_1.32.1      Rsamtools_2.12.0             </span></span>
<span><span class="co">## [125] GenomeInfoDbData_1.2.8        parallel_4.2.1               </span></span>
<span><span class="co">## [127] hms_1.1.2                     grid_4.2.1                   </span></span>
<span><span class="co">## [129] beachmat_2.12.0               rmarkdown_2.16               </span></span>
<span><span class="co">## [131] DelayedMatrixStats_1.18.0     shiny_1.7.2                  </span></span>
<span><span class="co">## [133] ggbeeswarm_0.6.0              restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephanie C. Hicks.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
