<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to Bioconductor • jhuquantneuro2022</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to Bioconductor">
<meta property="og:description" content="An introduction to Bioconductor
">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">jhuquantneuro2022</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/intro-bioconductor.html">Introduction to Bioconductor</a>
    </li>
    <li>
      <a href="../articles/intro-single-cell.html">Overview of single-cell analysis in R/Bioconductor</a>
    </li>
    <li>
      <a href="../articles/data-viz.html">Data visualizations and the tidyverse</a>
    </li>
    <li>
      <a href="../articles/problem-set.html">Problem Set 2</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://twitter.com/stephaniehicks" class="external-link">
    <span class="fa fa-twitter"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/stephaniehicks/jhuquantneuro2022" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to Bioconductor</h1>
                        <h4 data-toc-skip class="author">Stephanie
Hicks</h4>
            
      
      
      <div class="hidden name"><code>intro-bioconductor.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<div class="section level3">
<h3 id="key-resources">Key resources<a class="anchor" aria-label="anchor" href="#key-resources"></a>
</h3>
<ul>
<li>Workshop material: <a href="https://stephaniehicks.com/jhuquantneuro2022">pkgdown
website</a>
</li>
<li>Code: <a href="https://github.com/stephaniehicks/jhuquantneuro2022" class="external-link">GitHub</a>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Describe what is the Bioconductor and how it’s different than
CRAN</li>
<li>Describe the package types in Bioconductor</li>
<li>Recognize and work with core Bioconductor objects including
<code>SummarizedExperiment</code>
</li>
<li>Be able to perform a basic differential expression analysis with
bulk RNA-seq</li>
<li>Introduce the core Bioconductor object called
<code>SingleCellExperiment</code>.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="introduction-to-bioconductor">Introduction to Bioconductor<a class="anchor" aria-label="anchor" href="#introduction-to-bioconductor"></a>
</h2>
<div class="section level3">
<h3 id="a-brief-history-of-bioconductor">A brief history of Bioconductor<a class="anchor" aria-label="anchor" href="#a-brief-history-of-bioconductor"></a>
</h3>
<p>The Bioconductor project was <strong>started in the Fall of
2001</strong>, as an initiative for the collaborative creation of
extensible software for computational biology and bioinformatics.</p>
<p>The <strong>goal of the project</strong> is to develop tools for the
statistical analysis and comprehension of large datasets and
technological artifacts in rigorously and robustly designed experiments.
Beyond statistical analyses, the interpretation of statistical results
is supported by packages providing biological context, visualization,
and reproducibility.</p>
<p>Over the years, software packages contributed to the Bioconductor
project have reflected the evolution and emergence of several
high-throughput technologies, from microarrays to single-cell genomics,
through <strong>many variations of sequencing experiments</strong>
(e.g., RNA-seq, ChIP-seq, DNA-seq), analyses (e.g., sequence variation,
copy number variation, single nucleotide polymorphisms), and data
modalities (e.g., flow cytometry, proteomics, microscropy and image
analysis).</p>
<p>The Bioconductor project culminates at an <strong>annual conference
in North America in the summer</strong>, while regional conferences
offer great opportunities for networking in Europe, Asia, and North
America.</p>
<p>The project is <strong>committed to promote a diverse and inclusive
community</strong>, including a <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a> enforced by a Code of Conduct committee.</p>
<div class="figure">
<img src="https://github.com/carpentries-incubator/bioc-project/raw/main/fig/bioc-timeline.svg" alt=""><p class="caption">Timeline of major Bioconductor milestones alongside
technological advancements.</p>
</div>
<p><strong>Timeline of major Bioconductor milestones alongside
technological advancements.</strong> Above the timeline, the figure
marks the first occurrence of major events. Within the timeline, the
name of packages providing core infrastructure indicate the release
date. Below the timeline, major technological advancements contextualise
the evolution of the Bioconductor project over time.</p>
</div>
<div class="section level3">
<h3 id="a-scientific-project">A scientific project<a class="anchor" aria-label="anchor" href="#a-scientific-project"></a>
</h3>
<ul>
<li><p>The original publication describes the aims and methods of the
project at its inception is Gentleman et al. (2004).</p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590" class="external-link">Huber et
al. (2015)</a> illustrates the progression of the project, including
descriptions of core infrastructure and case studies, from the
perspective of both users and developers.</p></li>
<li><p><a href="https://pubmed.ncbi.nlm.nih.gov/31792435" class="external-link">Amezquita et
al. (2020)</a> reviewed further developments of the project in the wake
of single-cell genomics technologies.</p></li>
</ul>
<p>Many more publications and book chapters cite the Bioconductor
project, with recent example listed on the <a href="https://www.bioconductor.org/help/publications/" class="external-link">Bioconductor
website</a>.</p>
<p>In addition, there is a <strong>core team</strong> which is supported
by an NIH grant, and developers who contribute to the open source
Bioconductor packages.</p>
</div>
<div class="section level3">
<h3 id="a-package-repository">A package repository<a class="anchor" aria-label="anchor" href="#a-package-repository"></a>
</h3>
</div>
<div class="section level3">
<h3 id="overview-and-relationship-to-cran">Overview and relationship to CRAN<a class="anchor" aria-label="anchor" href="#overview-and-relationship-to-cran"></a>
</h3>
<p>Undoubtedly, <strong>software packages are the best-known aspect of
the Bioconductor project</strong>. Since its inception in 2001, the
repository has grown over time to host thousands of packages.</p>
<p>The Bioconductor project has extended the preexisting CRAN repository
– much larger and general-purpose in scope – to comprise R packages
primarily catering for bioinformatics and computational biology
analyses.</p>
</div>
<div class="section level3">
<h3 id="the-bioconductor-release-cycle">The Bioconductor release cycle<a class="anchor" aria-label="anchor" href="#the-bioconductor-release-cycle"></a>
</h3>
<p>The Bioconductor project also extended the packaging infrastructure
of the CRAN repository to better support the deployment and management
of packages at the user level.</p>
<p>In particular, the <strong>Bioconductor projects features a 6-month
release cycle</strong> (typically around April and October), which sees
a snapshot of the current version of all packages in the Bioconductor
repository earmarked for a specific version of R.</p>
<p><strong>R itself is released on an annual basis</strong> (typically
around April), meaning that for each release of R, two compatible
releases of Bioconductor packages are available.</p>
<ul>
<li><p>As such, <strong>Bioconductor package developers are required to
always use the version of R that will be associated with the next
release of the Bioconductor</strong> project.</p></li>
<li><p>This means using the development version of R between October and
April, and the release version of R between April and October.</p></li>
</ul>
<p>Why? The strict Bioconductor release cycle <strong>prevents users
from installing temporally distant versions of packages</strong> that
were very likely never tested together.</p>
<p>This practice <strong>reflects the development cycle of packages of
both CRAN and Bioconductor</strong>, where contemporaneous packages are
regularly tested by automated systems to ensure that the latest software
updates in package dependencies do not break downstream packages, or
prompts those package maintainers to update their own software as a
consequence.</p>
<p>Prior to each Bioconductor release, packages that do not pass the
requires suites of automated tests are deprecated and subsequently
removed from the repository.</p>
<p>This <strong>ensures that each Bioconductor release</strong> provides
a suite of packages that are mutually compatible, traceable, and
guaranteed to function for the associated version of R.</p>
<div class="figure">
<img src="https://github.com/carpentries-incubator/bioc-project/raw/main/fig/bioc-release-cycle.svg" alt=""><p class="caption">Timeline of release dates for selected Bioconductor
and R versions.</p>
</div>
<p><strong>Timeline of release dates for selected Bioconductor and R
versions.</strong> The upper section of the timeline indicates versions
and approximate release dates for the R project. The lower section of
the timeline indicates versions and release dates for the Bioconductor
project. Source: <a href="https://bioconductor.org/about/release-announcements/" class="external-link">Bioconductor</a>.</p>
</div>
<div class="section level3">
<h3 id="package-types">Package types<a class="anchor" aria-label="anchor" href="#package-types"></a>
</h3>
<p>Packages are broadly divided in four major categories:</p>
<ol style="list-style-type: decimal">
<li>software</li>
<li>annotation data</li>
<li>experiment data</li>
<li>workflows</li>
</ol>
<p><strong>Software packages</strong> themselves can be subdivided into
packages that</p>
<ul>
<li>provide <strong>infrastructure</strong> (i.e., classes) to store and
access data</li>
<li>packages that provide <strong>methodological tools</strong> to
process data stored in those data structures</li>
</ul>
<p>This <strong>separation of structure and analysis</strong> is at the
core of the Bioconductor project, encouraging developers of new
methodological software packages to thoughtfully re-use existing data
containers where possible, and <strong>reducing the cognitive
burden</strong> imposed on users who can more easily experiment with
alternative workflows without the need to learn and convert between
different data structures.</p>
<p><strong>Annotation data packages</strong> provide self-contained
databases of diverse genomic annotations (e.g., gene identifiers,
biological pathways).</p>
<ul>
<li>Different collections of annotation packages can be found in the
Bioconductor project.</li>
<li>They are identifiable by their respective naming pattern, and the
information that they contain.</li>
</ul>
<p>For instance, the so-called <code>OrgDb</code> packages (e.g., the
<em><a href="https://bioconductor.org/packages/3.15/org.Hs.eg.db" class="external-link">org.Hs.eg.db</a></em>
package) provide information mapping different types of gene identifiers
and pathway databases; the so-called <code>EnsDb</code> (e.g., <em><a href="https://bioconductor.org/packages/3.15/EnsDb.Hsapiens.v86" class="external-link">EnsDb.Hsapiens.v86</a></em>)
packages encapsulate individual versions of the Ensembl annotations in
Bioconductor packages; and the so-called <code>TxDb</code> packages
(e.g., <em><a href="https://bioconductor.org/packages/3.15/TxDb.Hsapiens.UCSC.hg38.knownGene" class="external-link">TxDb.Hsapiens.UCSC.hg38.knownGene</a></em>)
encapsulate individual versions UCSC gene annotation tables.</p>
<p><strong>Experiment data packages</strong> provide self-contained
datasets that are often used by software package developers to
demonstrate the use of their package on well-known standard datasets in
their <strong>package vignettes</strong>.</p>
<p>Finally, <strong>workflow packages</strong> exclusively provide
collections of vignettes that demonstrate the combined usage of several
other packages as a coherent workflow, but do not provide any new source
code or functionality themselves.</p>
</div>
<div class="section level3">
<h3 id="online-communication-channels">Online communication channels<a class="anchor" aria-label="anchor" href="#online-communication-channels"></a>
</h3>
<div class="section level4">
<h4 id="support-site">Support site<a class="anchor" aria-label="anchor" href="#support-site"></a>
</h4>
<p>The Bioconductor <a href="https://support.bioconductor.org/" class="external-link">support
site</a> provides a platform where users and developers can communicate
freely (following the Bioconductor <a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code of
Conduct</a>) to discuss issues on a range of subjects, ranging from
packages to conceptual questions about best practices.</p>
</div>
<div class="section level4">
<h4 id="slack-workspace">Slack workspace<a class="anchor" aria-label="anchor" href="#slack-workspace"></a>
</h4>
<p>The Bioconductor <a href="https://bioc-community.herokuapp.com/" class="external-link">Slack workspace</a> is an
open space that all community members are welcome to join (for free) and
use for rapid interactions. Currently, the “Pro” pricing plan kindly
supported by core funding provides:</p>
<ul>
<li>Unlimited message archive</li>
<li>Unlimited apps</li>
<li>Group video calls with screen sharing</li>
<li>Work securely with other organizations using Slack Connect</li>
</ul>
<p>A wide range of channels have been created to discuss a range of
subjects, and community members can freely join the discussion on those
channels of create new ones to discuss new subjects.</p>
<p>Important announcements are posted on the <code>#general</code>
channel.</p>
</div>
<div class="section level4">
<h4 id="developer-mailing-list">Developer Mailing List<a class="anchor" aria-label="anchor" href="#developer-mailing-list"></a>
</h4>
<p>The <a href="mailto:bioc-devel@r-project.org" class="email">bioc-devel@r-project.org</a> mailing list is used for
communication between package developers, and announcements from the
Biocondutor core team.</p>
</div>
<div class="section level4">
<h4 id="a-scientific-and-technical-community">A scientific and technical community<a class="anchor" aria-label="anchor" href="#a-scientific-and-technical-community"></a>
</h4>
<ul>
<li>
<a href="https://bioconductor.org/about/scientific-advisory-board/" class="external-link">Scientific
Advisory Board (SAB)</a> Meet Annually, External and Internal leader in
the field who act as project advisors. No Term limits.</li>
<li>
<a href="https://bioconductor.org/about/technical-advisory-board/" class="external-link">Technical
Advisory Board (TAB)</a>. Meet monthly to consider technical aspects of
core infastructure and scientific direction of the project. 15 members,
3 year term. Annual open-to-all elections to rotate members. Current
officers are Vince Carey (chair), Levi Waldron (vice Chair) Charlotte
Soneson (Secretary).</li>
<li>
<a href="https://bioconductor.org/about/community-advisory-board/" class="external-link">Community
Advisory Board (CAB)</a> Meet monthly to consider community outreach,
events, education and training. 15 members, 3 year term. Annual
open-to-all elections to rotate members. Current officers are Aedin
Culhane (chair), Matt Ritchie (co Chair), Lori Kern (Secretary).</li>
<li><a href="https://www.bioconductor.org/about/code-of-conduct/" class="external-link">Code
of Conduct committee</a></li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started-with-bioconductor">Getting started with Bioconductor<a class="anchor" aria-label="anchor" href="#getting-started-with-bioconductor"></a>
</h2>
<p>Before we get started, you need to know how to install Bioconductor
packages. The most important details are:</p>
<ul>
<li>Bioconductor is a package repository, like CRAN</li>
<li>All Bioconductor packages <strong>must</strong> be installed
following the instructions here: <a href="https://bioconductor.org/install" class="external-link uri">https://bioconductor.org/install</a>
</li>
<li>Bioconductor packages are linked in their versions, both to each
other and to the version of R</li>
<li>Bioconductor’s installation function will look up your version of R
and give you the appropriate versions of Bioconductor packages</li>
<li>If you want the latest version of Bioconductor, you need to use the
latest version of R.</li>
</ul>
<p>How do you know if a package is a Bioconductor package?</p>
<p>For one thing, you can just <strong>google the package name</strong>
and you will see either CRAN or Bioconductor as a first result (packages
must be in one or the other, they are not allowed to be on both
repositories).</p>
<p>But also, you can use Bioconductor’s installation function to install
any packages, even ones on CRAN.</p>
<p>By the way, you can install multiple packages at once by making a
string vector: <code>BiocManager::install(c("foo","bar"))</code></p>
<div class="section level3">
<h3 id="working-with-bioconductor-objects">Working with Bioconductor objects<a class="anchor" aria-label="anchor" href="#working-with-bioconductor-objects"></a>
</h3>
<p>Bioconductor’s infrastructure is <strong>built up of object
classes</strong>.</p>
<p>An example of a class is <strong>GRanges</strong> (stands for
“genomic ranges”), which is a way to specify a set of ranges in a
particular genome, e.g. from basepair 101 to basepair 200 on chromosome
1 of the human genome (version 38).</p>
<p>What’s an object?</p>
<p>Well <strong>everything in R is an object</strong>, but usually when
we talk about Bioconductor objects, we mean data structures containing
many attributes, so more complex than a vector or matrix.</p>
<p>And the objects have specific <strong>methods</strong> that help you
either access the information in the object, run analyses on the object,
plot the object, etc.</p>
<p>Bioconductor also allows for <strong>class hierarchy</strong>, which
means that you can define a class of object that inherits the structure
and methods of a <strong>superclass</strong> on which it depends. This
last point is mostly important for people who are developing new
software for Bioconductor (maybe that’s you!)</p>
<p>We will introduce the core Bioconductor objects here.</p>
<div class="section level4">
<h4 id="summarizedexperiment-se">SummarizedExperiment (SE)<a class="anchor" aria-label="anchor" href="#summarizedexperiment-se"></a>
</h4>
<p>First, we will discuss one of the most important classes of object,
which is the <em>SummarizedExperiment</em>, or SE.</p>
<p>SEs have the structure:</p>
<ul>
<li>a matrix of data, rows are genomic features, and columns are
samples</li>
<li>a table of data about the samples (columns)</li>
<li>a table of data about the features (rows)</li>
</ul>
<p>A diagram of this 3-part structure can be found <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4509590/figure/F2/" class="external-link">here</a>.</p>
<p>In SE, the 3 parts of the object are called 1) <code>assay</code>, 2)
<code>colData</code> and 3) <code>rowData</code> or
<code>rowRanges</code>.</p>
<p><strong>Note</strong>: There was a class of object that came before
the SE, called the <strong>ExpressionSet</strong>, which was used
primarily to store <strong>microarray</strong> data.</p>
<p>Here we will skip over the <strong>ExpressionSet</strong>, and just
look at SEs.</p>
<p>It helps to start by making a small toy SE, to see how the pieces
come together. (Often you won’t make an SE manually, but it will be
downloaded from an external source, or generated by a function that you
call, e.g. the <em><a href="https://bioconductor.org/packages/3.15/tximeta" class="external-link">tximeta</a></em>
or some other data loading function.)</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="va">col_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>sample<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">)</span>,</span>
<span>                       condition<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>,<span class="st">"A"</span>,<span class="st">"B"</span>,<span class="st">"B"</span>,<span class="st">"C"</span>,<span class="st">"C"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                       treated<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">1</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">col_data</span></span></code></pre></div>
<pre><code><span><span class="co">##   sample condition treated</span></span>
<span><span class="co">## 1      1         A       0</span></span>
<span><span class="co">## 2      2         A       1</span></span>
<span><span class="co">## 3      3         B       0</span></span>
<span><span class="co">## 4      4         B       1</span></span>
<span><span class="co">## 5      5         C       0</span></span>
<span><span class="co">## 6      6         C       1</span></span></code></pre>
<p>An important aspect of SEs is that the rows can optionally correspond
to particular set of <strong>GRanges</strong></p>
<ul>
<li>e.g. a row of an SE could give the number of RNA-seq reads that can
be assigned to a particular gene, and the row could also have metadata
in the 3rd slot including, e.g. location of the gene in the genome.</li>
</ul>
<p>In this case, we use the <code>rowRanges</code> slot to specify the
information.</p>
<p>If we don’t have ranges, we can just put a table on the “side” of the
SE by specifying <code>rowData</code>.</p>
<p>I will show in the example though how to provide
<code>rowRanges</code>.</p>
<p>Let’s <strong>use the first 10 genes in the Ensembl database</strong>
for human.</p>
<p>The following code loads a database, pulls out all the genes (as
<em>GRanges</em>), removes extra “non-standard” chromosomes, and then
subsets to the first 10 genes.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span></span>
<span><span class="va">g</span> <span class="op">&lt;-</span> <span class="fu">keepStandardChromosomes</span><span class="op">(</span><span class="va">g</span>, pruning.mode<span class="op">=</span><span class="st">"coarse"</span><span class="op">)</span></span>
<span><span class="va">row_ranges</span> <span class="op">&lt;-</span> <span class="va">g</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span></code></pre></div>
<p>We will make up some simulated “expression” measurements, and then
store these in the SE.</p>
<p>I call <code>list</code> so I can name the matrix, otherwise it would
not be named.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exprs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">6</span> <span class="op">*</span> <span class="fl">10</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fl">6</span>, nrow<span class="op">=</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/RangedSummarizedExperiment-class.html" class="external-link">SummarizedExperiment</a></span><span class="op">(</span>assay <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"exprs"</span> <span class="op">=</span> <span class="va">exprs</span><span class="op">)</span>,</span>
<span>                           colData <span class="op">=</span> <span class="va">col_data</span>,</span>
<span>                           rowRanges <span class="op">=</span> <span class="va">row_ranges</span><span class="op">)</span></span>
<span><span class="va">se</span></span></code></pre></div>
<pre><code><span><span class="co">## class: RangedSummarizedExperiment </span></span>
<span><span class="co">## dim: 10 6 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): exprs</span></span>
<span><span class="co">## rownames(10): ENSG00000223972 ENSG00000227232 ... ENSG00000238009</span></span>
<span><span class="co">##   ENSG00000239945</span></span>
<span><span class="co">## rowData names(6): gene_id gene_name ... symbol entrezid</span></span>
<span><span class="co">## colnames: NULL</span></span>
<span><span class="co">## colData names(3): sample condition treated</span></span></code></pre>
<p>We see this object has one named matrix. The object could have
multiple matrices (as long as these are the same shape).</p>
<p>In that case you could access the first with <code>assay</code> and
in general by name, e.g. <code>assay(se, "exprs")</code> or equivalently
<code>assays(se)[["exprs"]]</code> .</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "exprs"</span></span></code></pre>
<p>Finally, if we wanted to add data onto the rows, for example, the
score of a test on the matrix data, we use the metadata columns
function, or <code>mcols</code>:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">mcols</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">score</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 10 rows and 7 columns</span></span>
<span><span class="co">##                         gene_id    gene_name           gene_biotype</span></span>
<span><span class="co">##                     &lt;character&gt;  &lt;character&gt;            &lt;character&gt;</span></span>
<span><span class="co">## ENSG00000223972 ENSG00000223972      DDX11L1 transcribed_unproces..</span></span>
<span><span class="co">## ENSG00000227232 ENSG00000227232       WASH7P unprocessed_pseudogene</span></span>
<span><span class="co">## ENSG00000278267 ENSG00000278267    MIR6859-1                  miRNA</span></span>
<span><span class="co">## ENSG00000243485 ENSG00000243485    MIR1302-2                lincRNA</span></span>
<span><span class="co">## ENSG00000237613 ENSG00000237613      FAM138A                lincRNA</span></span>
<span><span class="co">## ENSG00000268020 ENSG00000268020       OR4G4P unprocessed_pseudogene</span></span>
<span><span class="co">## ENSG00000240361 ENSG00000240361      OR4G11P unprocessed_pseudogene</span></span>
<span><span class="co">## ENSG00000186092 ENSG00000186092        OR4F5         protein_coding</span></span>
<span><span class="co">## ENSG00000238009 ENSG00000238009 RP11-34P13.7                lincRNA</span></span>
<span><span class="co">## ENSG00000239945 ENSG00000239945 RP11-34P13.8                lincRNA</span></span>
<span><span class="co">##                 seq_coord_system       symbol                       entrezid</span></span>
<span><span class="co">##                      &lt;character&gt;  &lt;character&gt;                         &lt;list&gt;</span></span>
<span><span class="co">## ENSG00000223972       chromosome      DDX11L1 100287596,100287102,727856,...</span></span>
<span><span class="co">## ENSG00000227232       chromosome       WASH7P                             NA</span></span>
<span><span class="co">## ENSG00000278267       chromosome    MIR6859-1                      102466751</span></span>
<span><span class="co">## ENSG00000243485       chromosome    MIR1302-2            105376912,100302278</span></span>
<span><span class="co">## ENSG00000237613       chromosome      FAM138A           654835,645520,641702</span></span>
<span><span class="co">## ENSG00000268020       chromosome       OR4G4P                             NA</span></span>
<span><span class="co">## ENSG00000240361       chromosome      OR4G11P                             NA</span></span>
<span><span class="co">## ENSG00000186092       chromosome        OR4F5                          79501</span></span>
<span><span class="co">## ENSG00000238009       chromosome RP11-34P13.7                             NA</span></span>
<span><span class="co">## ENSG00000239945       chromosome RP11-34P13.8                             NA</span></span>
<span><span class="co">##                      score</span></span>
<span><span class="co">##                  &lt;numeric&gt;</span></span>
<span><span class="co">## ENSG00000223972  2.1701177</span></span>
<span><span class="co">## ENSG00000227232  0.0461614</span></span>
<span><span class="co">## ENSG00000278267  2.7315729</span></span>
<span><span class="co">## ENSG00000243485  0.2195109</span></span>
<span><span class="co">## ENSG00000237613  1.0484915</span></span>
<span><span class="co">## ENSG00000268020 -1.7197824</span></span>
<span><span class="co">## ENSG00000240361  0.2221783</span></span>
<span><span class="co">## ENSG00000186092 -0.5916436</span></span>
<span><span class="co">## ENSG00000238009 -1.4931525</span></span>
<span><span class="co">## ENSG00000239945 -2.4350805</span></span></code></pre>
<p>Adding data to the column metadata is even easier, we can just use
<code>$</code>:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span><span class="op">$</span><span class="va">librarySize</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">6</span>,<span class="fl">1e6</span>,<span class="fl">2e6</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##     sample condition  treated librarySize</span></span>
<span><span class="co">##   &lt;factor&gt;  &lt;factor&gt; &lt;factor&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">## 1        1         A        0     1940042</span></span>
<span><span class="co">## 2        2         A        1     1696274</span></span>
<span><span class="co">## 3        3         B        0     1799309</span></span>
<span><span class="co">## 4        4         B        1     1573100</span></span>
<span><span class="co">## 5        5         C        0     1102580</span></span>
<span><span class="co">## 6        6         C        1     1782852</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="using-the-ranges-of-a-se">Using the ranges of a SE<a class="anchor" aria-label="anchor" href="#using-the-ranges-of-a-se"></a>
</h4>
<p>How does this additional functionality of the <code>rowRanges</code>
facilitate faster data analysis?</p>
<p>Suppose we are working with another data set besides <code>se</code>
and we find a region of interest on chromsome 1.</p>
<p>If we want to pull out the expression data for that region, we just
ask for the subset of <code>se</code> that overlaps.</p>
<p>First, we <strong>build the query region</strong>, and then use the
<strong>GRanges</strong> function <code><a href="https://rdrr.io/pkg/IRanges/man/findOverlaps-methods.html" class="external-link">overlapsAny()</a></code> within
single square brackets (like you would subset any matrix-like
object):</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="fu">IRanges</span><span class="op">(</span><span class="fl">25000</span>,<span class="fl">40000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">overlapsAny</span><span class="op">(</span><span class="va">se</span>, <span class="va">query</span><span class="op">)</span>, <span class="op">]</span></span></code></pre></div>
<p>We could have equivalently used the shorthand code:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se_sub</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="va">se</span> <span class="op">%over%</span> <span class="va">query</span>,<span class="op">]</span></span></code></pre></div>
<p>We get just three ranges, and three rows of the SE:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 3 ranges and 7 metadata columns:</span></span>
<span><span class="co">##                   seqnames      ranges strand |         gene_id   gene_name</span></span>
<span><span class="co">##                      &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; |     &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   ENSG00000227232        1 14404-29570      - | ENSG00000227232      WASH7P</span></span>
<span><span class="co">##   ENSG00000243485        1 29554-31109      + | ENSG00000243485   MIR1302-2</span></span>
<span><span class="co">##   ENSG00000237613        1 34554-36081      - | ENSG00000237613     FAM138A</span></span>
<span><span class="co">##                             gene_biotype seq_coord_system      symbol</span></span>
<span><span class="co">##                              &lt;character&gt;      &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   ENSG00000227232 unprocessed_pseudogene       chromosome      WASH7P</span></span>
<span><span class="co">##   ENSG00000243485                lincRNA       chromosome   MIR1302-2</span></span>
<span><span class="co">##   ENSG00000237613                lincRNA       chromosome     FAM138A</span></span>
<span><span class="co">##                               entrezid     score</span></span>
<span><span class="co">##                                 &lt;list&gt; &lt;numeric&gt;</span></span>
<span><span class="co">##   ENSG00000227232                 &lt;NA&gt; 0.0461614</span></span>
<span><span class="co">##   ENSG00000243485  105376912,100302278 0.2195109</span></span>
<span><span class="co">##   ENSG00000237613 654835,645520,641702 1.0484915</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from GRCh38 genome</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">se_sub</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                        [,1]      [,2]       [,3]      [,4]       [,5]</span></span>
<span><span class="co">## ENSG00000227232 -0.08079092 -1.585902 -0.5836049 0.7321954 -0.9666263</span></span>
<span><span class="co">## ENSG00000243485 -1.29885760  1.298616 -0.2868365 0.2591491 -1.6483581</span></span>
<span><span class="co">## ENSG00000237613  0.84994374 -1.451339 -1.7064174 1.2807216  0.2915711</span></span>
<span><span class="co">##                       [,6]</span></span>
<span><span class="co">## ENSG00000227232 -2.2637416</span></span>
<span><span class="co">## ENSG00000243485 -0.3265077</span></span>
<span><span class="co">## ENSG00000237613  0.1643328</span></span></code></pre>
<p>Another useful property is that we know metadata about the
chromosomes, and the version of the genome.</p>
<ul>
<li>
<strong>Note</strong>: If you were not yet aware, the basepair
position of a given feature, say gene <em>XYZ</em>, will change between
versions of the genome, as sequences are added or rearranged.</li>
</ul>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Seqinfo object with 25 sequences (1 circular) from GRCh38 genome:</span></span>
<span><span class="co">##   seqnames seqlengths isCircular genome</span></span>
<span><span class="co">##   1         248956422      FALSE GRCh38</span></span>
<span><span class="co">##   10        133797422      FALSE GRCh38</span></span>
<span><span class="co">##   11        135086622      FALSE GRCh38</span></span>
<span><span class="co">##   12        133275309      FALSE GRCh38</span></span>
<span><span class="co">##   13        114364328      FALSE GRCh38</span></span>
<span><span class="co">##   ...             ...        ...    ...</span></span>
<span><span class="co">##   8         145138636      FALSE GRCh38</span></span>
<span><span class="co">##   9         138394717      FALSE GRCh38</span></span>
<span><span class="co">##   MT            16569       TRUE GRCh38</span></span>
<span><span class="co">##   X         156040895      FALSE GRCh38</span></span>
<span><span class="co">##   Y          57227415      FALSE GRCh38</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="downloading-se-data">Downloading SE data<a class="anchor" aria-label="anchor" href="#downloading-se-data"></a>
</h3>
<p>Let’s download a SE object from <a href="https://jhubiostatistics.shinyapps.io/recount/" class="external-link">recount2</a>,
which performs a basic summarization of public data sets with gene
expression data.</p>
<p>This <strong>dataset contains RNA-seq samples from human airway
epithelial cell cultures</strong>.</p>
<p>The paper is <a href="https://www.ncbi.nlm.nih.gov/pubmed/25706956" class="external-link">here</a>. The
structure of the experiment was that, cell cultures from 6 asthmatic and
6 non-asthmatics donors were treated with viral infection or left
untreated (controls).</p>
<p>So we have 2 samples (control or treated) for each of the 12
donors.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://here.r-lib.org/" class="external-link">here</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># tests if a directory named "data" exists locally</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span> <span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span><span class="op">)</span><span class="op">)</span> <span class="op">}</span></span>
<span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="st">"asthma.rda"</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="va">file</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">url</span> <span class="op">&lt;-</span> <span class="st">"http://duffel.rail.bio/recount/SRP046226/rse_gene.Rdata"</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">url</span>, <span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="va">file</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html" class="external-link">here</a></span><span class="op">(</span><span class="st">"data"</span>, <span class="va">file</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We use a custom function to produce a matrix which a count of RNA
fragments for each gene (rows) and each sample (columns).</p>
<p>(Recount project calls these objects <code>rse</code> for
<strong>RangedSummarizedExperiment</strong>, meaning it has
<code>rowRanges</code> information.)</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_scale_counts</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rse_gene</span>, <span class="va">round</span><span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">cts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span></span>
<span>  <span class="co"># mapped_read_count includes the count for both reads of a pair</span></span>
<span>  <span class="va">paired</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">paired_end</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span></span>
<span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">mapped_read_count</span> <span class="op">/</span> <span class="va">paired</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="va">cts</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span> <span class="op">*</span> <span class="va">x</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">round</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">rse_gene</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">rse</span> <span class="op">&lt;-</span> <span class="fu">my_scale_counts</span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span></span></code></pre></div>
<p>We can take a peek at the column data:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 24 rows and 6 columns</span></span>
<span><span class="co">##                project      sample  experiment         run</span></span>
<span><span class="co">##            &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## SRR1565926   SRP046226   SRS694613   SRX692912  SRR1565926</span></span>
<span><span class="co">## SRR1565927   SRP046226   SRS694614   SRX692913  SRR1565927</span></span>
<span><span class="co">## SRR1565928   SRP046226   SRS694615   SRX692914  SRR1565928</span></span>
<span><span class="co">## SRR1565929   SRP046226   SRS694616   SRX692915  SRR1565929</span></span>
<span><span class="co">## SRR1565930   SRP046226   SRS694617   SRX692916  SRR1565930</span></span>
<span><span class="co">## ...                ...         ...         ...         ...</span></span>
<span><span class="co">## SRR1565945   SRP046226   SRS694632   SRX692931  SRR1565945</span></span>
<span><span class="co">## SRR1565946   SRP046226   SRS694633   SRX692932  SRR1565946</span></span>
<span><span class="co">## SRR1565947   SRP046226   SRS694634   SRX692933  SRR1565947</span></span>
<span><span class="co">## SRR1565948   SRP046226   SRS694635   SRX692934  SRR1565948</span></span>
<span><span class="co">## SRR1565949   SRP046226   SRS694636   SRX692935  SRR1565949</span></span>
<span><span class="co">##            read_count_as_reported_by_sra reads_downloaded</span></span>
<span><span class="co">##                                &lt;integer&gt;        &lt;integer&gt;</span></span>
<span><span class="co">## SRR1565926                      12866750         12866750</span></span>
<span><span class="co">## SRR1565927                      12797108         12797108</span></span>
<span><span class="co">## SRR1565928                      13319016         13319016</span></span>
<span><span class="co">## SRR1565929                      13725752         13725752</span></span>
<span><span class="co">## SRR1565930                      10882416         10882416</span></span>
<span><span class="co">## ...                                  ...              ...</span></span>
<span><span class="co">## SRR1565945                      13791854         13791854</span></span>
<span><span class="co">## SRR1565946                      13480842         13480842</span></span>
<span><span class="co">## SRR1565947                      13166594         13166594</span></span>
<span><span class="co">## SRR1565948                      13320398         13320398</span></span>
<span><span class="co">## SRR1565949                      13002276         13002276</span></span></code></pre>
<p>The information we are interested in is contained in the
<code>characteristics</code> column (which is a character list).</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">characteristics</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "CompressedCharacterList"</span></span>
<span><span class="co">## attr(,"package")</span></span>
<span><span class="co">## [1] "IRanges"</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span><span class="op">$</span><span class="va">characteristics</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## CharacterList of length 3</span></span>
<span><span class="co">## [[1]] cell type: Isolated from human trachea-bronchial tissues ...</span></span>
<span><span class="co">## [[2]] cell type: Isolated from human trachea-bronchial tissues ...</span></span>
<span><span class="co">## [[3]] cell type: Isolated from human trachea-bronchial tissues ...</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span><span class="op">$</span><span class="va">characteristics</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "cell type: Isolated from human trachea-bronchial tissues"</span></span>
<span><span class="co">## [2] "passages: 2"                                             </span></span>
<span><span class="co">## [3] "disease state: asthmatic"                                </span></span>
<span><span class="co">## [4] "treatment: HRV16"</span></span></code></pre>
<p>We can pull out the 3 and 4 element using the <code>sapply</code>
function and the square bracket function.</p>
<p>I know this syntax looks a little funny, but it’s really just saying,
use the single square bracket, pull out the third element (or fourth
element).</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rse</span><span class="op">$</span><span class="va">condition</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">characteristics</span>, <span class="va">`[`</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="va">rse</span><span class="op">$</span><span class="va">treatment</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">characteristics</span>, <span class="va">`[`</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">condition</span>, <span class="va">rse</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                               </span></span>
<span><span class="co">##                                treatment: HRV16 treatment: Vehicle</span></span>
<span><span class="co">##   disease state: asthmatic                    6                  6</span></span>
<span><span class="co">##   disease state: non-asthmatic                6                  6</span></span></code></pre>
<p>Let’s see what the <code>rowRanges</code> of this experiment look
like:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowRanges.html" class="external-link">rowRanges</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 58037 ranges and 3 metadata columns:</span></span>
<span><span class="co">##                      seqnames              ranges strand |            gene_id</span></span>
<span><span class="co">##                         &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |        &lt;character&gt;</span></span>
<span><span class="co">##   ENSG00000000003.14     chrX 100627109-100639991      - | ENSG00000000003.14</span></span>
<span><span class="co">##    ENSG00000000005.5     chrX 100584802-100599885      + |  ENSG00000000005.5</span></span>
<span><span class="co">##   ENSG00000000419.12    chr20   50934867-50958555      - | ENSG00000000419.12</span></span>
<span><span class="co">##   ENSG00000000457.13     chr1 169849631-169894267      - | ENSG00000000457.13</span></span>
<span><span class="co">##   ENSG00000000460.16     chr1 169662007-169854080      + | ENSG00000000460.16</span></span>
<span><span class="co">##                  ...      ...                 ...    ... .                ...</span></span>
<span><span class="co">##    ENSG00000283695.1    chr19   52865369-52865429      - |  ENSG00000283695.1</span></span>
<span><span class="co">##    ENSG00000283696.1     chr1 161399409-161422424      + |  ENSG00000283696.1</span></span>
<span><span class="co">##    ENSG00000283697.1     chrX 149548210-149549852      - |  ENSG00000283697.1</span></span>
<span><span class="co">##    ENSG00000283698.1     chr2 112439312-112469687      - |  ENSG00000283698.1</span></span>
<span><span class="co">##    ENSG00000283699.1    chr10   12653138-12653197      - |  ENSG00000283699.1</span></span>
<span><span class="co">##                      bp_length          symbol</span></span>
<span><span class="co">##                      &lt;integer&gt; &lt;CharacterList&gt;</span></span>
<span><span class="co">##   ENSG00000000003.14      4535          TSPAN6</span></span>
<span><span class="co">##    ENSG00000000005.5      1610            TNMD</span></span>
<span><span class="co">##   ENSG00000000419.12      1207            DPM1</span></span>
<span><span class="co">##   ENSG00000000457.13      6883           SCYL3</span></span>
<span><span class="co">##   ENSG00000000460.16      5967        C1orf112</span></span>
<span><span class="co">##                  ...       ...             ...</span></span>
<span><span class="co">##    ENSG00000283695.1        61            &lt;NA&gt;</span></span>
<span><span class="co">##    ENSG00000283696.1       997            &lt;NA&gt;</span></span>
<span><span class="co">##    ENSG00000283697.1      1184    LOC101928917</span></span>
<span><span class="co">##    ENSG00000283698.1       940            &lt;NA&gt;</span></span>
<span><span class="co">##    ENSG00000283699.1        60         MIR4481</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 25 sequences (1 circular) from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/GenomeInfoDb/man/seqinfo.html" class="external-link">seqinfo</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Seqinfo object with 25 sequences (1 circular) from an unspecified genome; no seqlengths:</span></span>
<span><span class="co">##   seqnames seqlengths isCircular genome</span></span>
<span><span class="co">##   chr1           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chr2           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chr3           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chr4           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chr5           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   ...             ...        ...    ...</span></span>
<span><span class="co">##   chr21          &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chr22          &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chrX           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chrY           &lt;NA&gt;       &lt;NA&gt;   &lt;NA&gt;</span></span>
<span><span class="co">##   chrM           &lt;NA&gt;       TRUE   &lt;NA&gt;</span></span></code></pre>
<p>The <code>rowRanges</code> here were determined by the quantification
method that the <em>recount2</em> authors used.</p>
<p>We don’t know what the genome is from the <code>seqinfo</code>, but
we could look this up from the project website.</p>
<p>The following code I use to clean up the condition and treatment
variables:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="va">rse</span><span class="op">$</span><span class="va">condition</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"-"</span>,<span class="st">"."</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"disease state: (.*)"</span>,<span class="st">"\\1"</span>,<span class="va">x</span><span class="op">)</span> <span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span><span class="op">)</span></span>
<span><span class="va">rse</span><span class="op">$</span><span class="va">treatment</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"treatment: (.*)"</span>,<span class="st">"\\1"</span>,<span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we have:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">rse</span><span class="op">$</span><span class="va">condition</span>, <span class="va">rse</span><span class="op">$</span><span class="va">treatment</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                </span></span>
<span><span class="co">##                 HRV16 Vehicle</span></span>
<span><span class="co">##   asthmatic         6       6</span></span>
<span><span class="co">##   non.asthmatic     6       6</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualizing-count-matrix-data-in-a-se">Visualizing count matrix data in a SE<a class="anchor" aria-label="anchor" href="#visualizing-count-matrix-data-in-a-se"></a>
</h3>
<p>Here we just use a transformation so that we can compute meaningful
distances on count data (without a larger discussion on
normalization).</p>
<p>We build a <strong>DESeqDataSet</strong> and then specify the
experimental design using a <code>~</code> and the variables that we
expect to produce differences in the counts. (These variables are used
to assess how much technical variability is in the data, but not used in
the transformation function itself.)</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSet</a></span><span class="op">(</span><span class="va">rse</span>, <span class="op">~</span><span class="va">condition</span> <span class="op">+</span> <span class="va">treatment</span><span class="op">)</span></span></code></pre></div>
<p>We use this function, which implements a <strong>variance stabilizing
transformation</strong>:</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/DESeq2/man/vst.html" class="external-link">vst</a></span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre></div>
<p>We calculate the variance across all samples (on the transformed
data):</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/HenrikBengtsson/matrixStats" class="external-link">matrixStats</a></span><span class="op">)</span></span>
<span><span class="va">rv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/matrixStats/man/rowVars.html" class="external-link">rowVars</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">o</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">rv</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">100</span><span class="op">]</span></span></code></pre></div>
<p>Finally, before plotting a heatmap, we <strong>extract the
covariates</strong> that we want to annotated the top of the plot.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">anno_col</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"condition"</span>,<span class="st">"treatment"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">anno_col</span></span></code></pre></div>
<pre><code><span><span class="co">##                condition treatment</span></span>
<span><span class="co">## SRR1565926     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565927     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565928     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565929     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565930     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565931     asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565932     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565933     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565934     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565935     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565936     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565937     asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565938 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565939 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565940 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565941 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565942 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565943 non.asthmatic     HRV16</span></span>
<span><span class="co">## SRR1565944 non.asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565945 non.asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565946 non.asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565947 non.asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565948 non.asthmatic   Vehicle</span></span>
<span><span class="co">## SRR1565949 non.asthmatic   Vehicle</span></span></code></pre>
<p>This code pull out the top of the transformed data by variance, and
adds an annotation to the top of the plot.</p>
<p>By default the rows and columns will be clustered by Euclidean
distance. See <code><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">?pheatmap</a></code> for more details on this function
(it’s a very detailed manual page).</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">o</span>,<span class="op">]</span>,</span>
<span>         annotation_col<span class="op">=</span><span class="va">anno_col</span>,</span>
<span>         show_rownames<span class="op">=</span><span class="cn">FALSE</span>, </span>
<span>         show_colnames<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro-bioconductor_files/figure-html/unnamed-chunk-23-1.png" width="700"></p>
<p>We can also easily make a PCA plot with dedicated functions:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/plotPCA.html" class="external-link">plotPCA</a></span><span class="op">(</span><span class="va">vsd</span>, intgroup<span class="op">=</span><span class="st">"treatment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro-bioconductor_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="singlecellexperiment">SingleCellExperiment<a class="anchor" aria-label="anchor" href="#singlecellexperiment"></a>
</h2>
<p>An example of a class that extends the SE is
<strong>SingleCellExperiment</strong>. This is a special object type for
looking at single cell data.</p>
<p>For more details, there is a free online book “Orchestrating Single
Cell Analysis With Bioconductor” produced by a group within the
Bioconductor Project, with lots of example analyses: <a href="http://bioconductor.org/books/3.15/OSCA.basic/" class="external-link">OSCA</a>.</p>
<p>Here we show a quick example of how this object extends the SE.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">rse</span>, <span class="st">"SingleCellExperiment"</span><span class="op">)</span></span>
<span><span class="va">sce</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SingleCellExperiment </span></span>
<span><span class="co">## dim: 58037 24 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(58037): ENSG00000000003.14 ENSG00000000005.5 ...</span></span>
<span><span class="co">##   ENSG00000283698.1 ENSG00000283699.1</span></span>
<span><span class="co">## rowData names(3): gene_id bp_length symbol</span></span>
<span><span class="co">## colnames(24): SRR1565926 SRR1565927 ... SRR1565948 SRR1565949</span></span>
<span><span class="co">## colData names(23): project sample ... condition treatment</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span></code></pre>
<p>There are special functions dedicated to scaling the samples (we will
discuss this technical aspect soon):</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">sizeFactors</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] 0.7672143 0.8205514 0.8686567 0.9479224 0.6484723 0.9815079 1.0797070</span></span>
<span><span class="co">##  [8] 1.0569889 1.4377886 0.9465292 1.4759422 1.2630195 0.8889808 1.0524670</span></span>
<span><span class="co">## [15] 0.9677885 0.8086102 0.8806503 0.8999780 0.9505805 1.0430322 1.2527967</span></span>
<span><span class="co">## [22] 0.9908707 0.5208294 1.4491155</span></span></code></pre>
<p>Similarly, dedicated functions for transformations:</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assayNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "counts"    "logcounts"</span></span></code></pre>
<p>And dedicated functions and new slots for reduced dimensions:</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html" class="external-link">fixedPCA</a></span><span class="op">(</span><span class="va">sce</span>, rank<span class="op">=</span><span class="fl">5</span>, subset.row<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDimNames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "PCA"</span></span></code></pre>
<p>We can manually get at the PCs:</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pca</span><span class="op">[</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro-bioconductor_files/figure-html/unnamed-chunk-29-1.png" width="700"></p>
<p>But we can more easily use dedicated visualization functions:</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">sce</span>, <span class="st">"PCA"</span>, colour_by<span class="op">=</span><span class="st">"treatment"</span><span class="op">)</span></span></code></pre></div>
<p><img src="intro-bioconductor_files/figure-html/unnamed-chunk-30-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session Info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin21.6.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Monterey 12.5.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /opt/homebrew/Cellar/openblas/0.3.21/lib/libopenblasp-r0.3.21.dylib</span></span>
<span><span class="co">## LAPACK: /opt/homebrew/Cellar/r/4.2.1_2/lib/R/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] scater_1.24.0               ggplot2_3.3.6              </span></span>
<span><span class="co">##  [3] scran_1.24.0                scuttle_1.6.3              </span></span>
<span><span class="co">##  [5] SingleCellExperiment_1.18.0 pheatmap_1.0.12            </span></span>
<span><span class="co">##  [7] DESeq2_1.36.0               magrittr_2.0.3             </span></span>
<span><span class="co">##  [9] here_1.0.1                  EnsDb.Hsapiens.v86_2.99.0  </span></span>
<span><span class="co">## [11] ensembldb_2.20.2            AnnotationFilter_1.20.0    </span></span>
<span><span class="co">## [13] GenomicFeatures_1.48.3      AnnotationDbi_1.58.0       </span></span>
<span><span class="co">## [15] SummarizedExperiment_1.26.1 Biobase_2.56.0             </span></span>
<span><span class="co">## [17] GenomicRanges_1.48.0        GenomeInfoDb_1.32.3        </span></span>
<span><span class="co">## [19] IRanges_2.30.1              S4Vectors_0.34.0           </span></span>
<span><span class="co">## [21] BiocGenerics_0.42.0         MatrixGenerics_1.8.1       </span></span>
<span><span class="co">## [23] matrixStats_0.62.0         </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] BiocFileCache_2.4.0       systemfonts_1.0.4        </span></span>
<span><span class="co">##   [3] igraph_1.3.4              lazyeval_0.2.2           </span></span>
<span><span class="co">##   [5] splines_4.2.1             BiocParallel_1.30.3      </span></span>
<span><span class="co">##   [7] digest_0.6.29             htmltools_0.5.3          </span></span>
<span><span class="co">##   [9] viridis_0.6.2             fansi_1.0.3              </span></span>
<span><span class="co">##  [11] memoise_2.0.1             ScaledMatrix_1.4.0       </span></span>
<span><span class="co">##  [13] cluster_2.1.4             limma_3.52.2             </span></span>
<span><span class="co">##  [15] Biostrings_2.64.1         annotate_1.74.0          </span></span>
<span><span class="co">##  [17] pkgdown_2.0.6             prettyunits_1.1.1        </span></span>
<span><span class="co">##  [19] colorspace_2.0-3          ggrepel_0.9.1            </span></span>
<span><span class="co">##  [21] blob_1.2.3                rappdirs_0.3.3           </span></span>
<span><span class="co">##  [23] textshaping_0.3.6         xfun_0.32                </span></span>
<span><span class="co">##  [25] dplyr_1.0.10              crayon_1.5.1             </span></span>
<span><span class="co">##  [27] RCurl_1.98-1.8            jsonlite_1.8.0           </span></span>
<span><span class="co">##  [29] genefilter_1.78.0         survival_3.4-0           </span></span>
<span><span class="co">##  [31] glue_1.6.2                gtable_0.3.1             </span></span>
<span><span class="co">##  [33] zlibbioc_1.42.0           XVector_0.36.0           </span></span>
<span><span class="co">##  [35] DelayedArray_0.22.0       BiocSingular_1.12.0      </span></span>
<span><span class="co">##  [37] scales_1.2.1              DBI_1.1.3                </span></span>
<span><span class="co">##  [39] edgeR_3.38.4              Rcpp_1.0.9               </span></span>
<span><span class="co">##  [41] viridisLite_0.4.1         xtable_1.8-4             </span></span>
<span><span class="co">##  [43] progress_1.2.2            dqrng_0.3.0              </span></span>
<span><span class="co">##  [45] bit_4.0.4                 rsvd_1.0.5               </span></span>
<span><span class="co">##  [47] metapod_1.4.0             httr_1.4.4               </span></span>
<span><span class="co">##  [49] RColorBrewer_1.1-3        ellipsis_0.3.2           </span></span>
<span><span class="co">##  [51] pkgconfig_2.0.3           XML_3.99-0.10            </span></span>
<span><span class="co">##  [53] farver_2.1.1              sass_0.4.2               </span></span>
<span><span class="co">##  [55] dbplyr_2.2.1              locfit_1.5-9.6           </span></span>
<span><span class="co">##  [57] utf8_1.2.2                tidyselect_1.1.2         </span></span>
<span><span class="co">##  [59] labeling_0.4.2            rlang_1.0.5              </span></span>
<span><span class="co">##  [61] munsell_0.5.0             tools_4.2.1              </span></span>
<span><span class="co">##  [63] cachem_1.0.6              cli_3.3.0                </span></span>
<span><span class="co">##  [65] generics_0.1.3            RSQLite_2.2.16           </span></span>
<span><span class="co">##  [67] evaluate_0.16             stringr_1.4.1            </span></span>
<span><span class="co">##  [69] fastmap_1.1.0             yaml_2.3.5               </span></span>
<span><span class="co">##  [71] ragg_1.2.2                knitr_1.40               </span></span>
<span><span class="co">##  [73] bit64_4.0.5               fs_1.5.2                 </span></span>
<span><span class="co">##  [75] purrr_0.3.4               KEGGREST_1.36.3          </span></span>
<span><span class="co">##  [77] sparseMatrixStats_1.8.0   xml2_1.3.3               </span></span>
<span><span class="co">##  [79] biomaRt_2.52.0            BiocStyle_2.24.0         </span></span>
<span><span class="co">##  [81] compiler_4.2.1            rstudioapi_0.14          </span></span>
<span><span class="co">##  [83] beeswarm_0.4.0            filelock_1.0.2           </span></span>
<span><span class="co">##  [85] curl_4.3.2                png_0.1-7                </span></span>
<span><span class="co">##  [87] statmod_1.4.37            tibble_3.1.8             </span></span>
<span><span class="co">##  [89] geneplotter_1.74.0        bslib_0.4.0              </span></span>
<span><span class="co">##  [91] stringi_1.7.8             highr_0.9                </span></span>
<span><span class="co">##  [93] desc_1.4.1                lattice_0.20-45          </span></span>
<span><span class="co">##  [95] bluster_1.6.0             ProtGenerics_1.28.0      </span></span>
<span><span class="co">##  [97] Matrix_1.4-1              vctrs_0.4.1              </span></span>
<span><span class="co">##  [99] pillar_1.8.1              lifecycle_1.0.1          </span></span>
<span><span class="co">## [101] BiocManager_1.30.18       jquerylib_0.1.4          </span></span>
<span><span class="co">## [103] BiocNeighbors_1.14.0      bitops_1.0-7             </span></span>
<span><span class="co">## [105] irlba_2.3.5               rtracklayer_1.56.1       </span></span>
<span><span class="co">## [107] R6_2.5.1                  BiocIO_1.6.0             </span></span>
<span><span class="co">## [109] gridExtra_2.3             vipor_0.4.5              </span></span>
<span><span class="co">## [111] codetools_0.2-18          assertthat_0.2.1         </span></span>
<span><span class="co">## [113] rprojroot_2.0.3           rjson_0.2.21             </span></span>
<span><span class="co">## [115] withr_2.5.0               GenomicAlignments_1.32.1 </span></span>
<span><span class="co">## [117] Rsamtools_2.12.0          GenomeInfoDbData_1.2.8   </span></span>
<span><span class="co">## [119] parallel_4.2.1            hms_1.1.2                </span></span>
<span><span class="co">## [121] grid_4.2.1                beachmat_2.12.0          </span></span>
<span><span class="co">## [123] rmarkdown_2.16            DelayedMatrixStats_1.18.0</span></span>
<span><span class="co">## [125] ggbeeswarm_0.6.0          restfulr_0.0.15</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stephanie C. Hicks.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
